--- 
title: "The intrinsic and extrinsic effects of Tet proteins during gastrulation"
author: "Markus Mittnenzweig & Aviezer Lifshitz"
date: "`r Sys.Date()`"
knit: "bookdown::render_book"
site: bookdown::gitbook_site
description: "Analysis and code for the paper: The intrinsic and extrinsic effects of Tet proteins during gastrulation"
url: 'tanaylab.github.io/tet-gastrulation'
github-repo: tanaylab/tet-gastrulation
---

Following is the code that generates the figures of our paper on TET function during gastrulation. The code is splitted to jupyter notebooks that can be found at: https://github.com/tanaylab/tet-gastrulation

## Run the notebooks

Prior to any analysis, after cloning the repository, please download first the necessary data by running (in the root directory of the cloned repository):


```bash
R -e "source('scripts/download_data.R'); download_full_data()"
```

This will download all the necessary data including processed metacell objects necessary for generating the figures. You can also download a minimal version ( `download_minimal_scrna_data()` ) in which case you would need to rerun all the notebooks prior to _generate-figures_ (see section *Notebook order* below.)

## Necessary R packages

You can install the necessary R packages by running:

```r
remotes::install_github("tanaylab/tet-gastrulation", subdir = "scripts/tet.gastru")
```

The initialization script (`scripts/init.R`) loads automatically the necessary R packages to run the notebooks. The analysis was done using R 4.0.5 and the following packages:

- gridExtra 2.3
- qvalue 2.22.0
- ggrepel 0.9.1
- ggpubr 0.4.0   
- ggplot2 3.3.5
- zoo 1.8-9
- qlcMatrix 0.9.7
- sparsesvd 0.2  
- slam 0.1-49
- tidyr 1.2.0
- dplyr 1.0.9
- tgutil 0.1.13  
- tgstat 2.3.17
- metacell 0.3.7
- Matrix 1.3-4

## Notebook order 

You should run the notebooks _import_plates_ and _gating_ first and _generate-figures_ last.

1. import_plates
2. gating
3. TKO-chimera-embryos-processing
4. TKO-tetraploid-complemented-embryos-processing
5. Control-tetraploid-complemented-embryos-processing
6. DKO-chimera-embryos-processing
7. AAV_Cre-TKO-embryos-processing
8. Methylation
9. generate-figures


<!--chapter:end:index.Rmd-->

---
jupyter:
  jupytext:
    formats: ipynb,Rmd
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.4
  kernelspec:
    display_name: R
    language: R
    name: ir353
---

```{r, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
setwd(here::here())
```

```{r}
library("Matrix")
library("dplyr")
library("metacell")
library("tgstat")
```

```{r}
scdb_init("scrna_db", force_reinit = T)
```

```{r}
mat <- scdb_mat("tko_germline")
```

```{r}
# for compatibility with later scripts, change column name 'cell_genotype' to 'cell_type'
md <- mat@cell_metadata

if ("cell_genotype" %in% colnames(md)) {
    md <- rename(md, cell_type = cell_genotype)
}

mat@cell_metadata <- md
scdb_add_mat(id = "tko_germline", mat = mat)
```

#### Remove extraembryonic ectoderm and parietal endoderm cells
#### Merge cells with the wildtype atlas

```{r}
source("scripts/pipeline/remove_exe_ectoderm_and_parietal_endo_cls.R")
source("scripts/pipeline/merge_umi_mat_with_wt10_umi_mat.R")

mat <- remove_exe_ectoderm_and_parietal_endoderm(mat_query = mat)
mat_new <- merge_umi_mat_with_wt10(scmat = mat)
scdb_add_mat(id = "tko_germline_wt10", mat = mat_new)
```

#### Generate single-cell balanced kNN-graph
This step needs the file  
_data/tet_tko.bad_genes.txt_

```{r}
# 4. generate cgraph
source("scripts/pipeline/gen_cgraph.R")
gen_cgraph(mat_nm = "tko_germline_wt10", force_recompute_cgraph = F)
```

#### Transfer cell type annotation to TKO cells

```{r}
source("scripts/pipeline/transfer_cell_type_annotation.R")
transfer_color_chimera_tetraploid("tko_germline_wt10")
```

#### Time single embryos

```{r}
source("scripts/pipeline/transfer_time_annotation.R")
source("scripts/pipeline/tko_germline_timing.R")
tko_germline_timing("tko_germline_wt10")
```

```{r}

```

<!--chapter:end:AAV_Cre-TKO-embryos-processing.Rmd-->

---
jupyter:
  jupytext:
    formats: ipynb,Rmd
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.4
  kernelspec:
    display_name: R
    language: R
    name: ir353
---

```{r, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
setwd(here::here())
```

```{r}
library("metacell")
library("tgstat")
library("tgconfig")
library("Matrix")
library("dplyr")
```

```{r}
scdb_init("scrna_db/", force_reinit = T)
```

```{r}
mat1 <- scdb_mat("control_tetra")
```

#### Remove extraembryonic ectoderm and parietal endoderm cells

```{r}
source("scripts/pipeline/remove_exe_ectoderm_and_parietal_endo_cls.R")
mat1 <- remove_exe_ectoderm_and_parietal_endoderm(mat_query = mat1)
```

#### Merge cells with the wildtype atlas
We first merge this matrix with the tetraploid-complemented control embryos from Mittnenzweig, Mayshar et al. Cell (2021)

```{r}
source("scripts/pipeline/preprocessing_foxc_experiment_control_tetraploid_mat.R")
preprocessing_tetraploid_control_mat_from_embflow_paper()
```

```{r}
source("scripts/pipeline/merge_umi_mat_with_wt10_umi_mat.R")
```

```{r}
source("scripts/pipeline/merge_umi_mat_with_wt10_umi_mat.R")
mat2 <- scdb_mat("control_tetra_from_embflow_paper_f")

mat_new <- merge_two_scmat(mat1 = mat1, mat2 = mat2)

scdb_add_mat("control_tetra_all", mat = mat_new)
```

Next merge matrix with WT cells

```{r}
mat_new <- scdb_mat("control_tetra_all")
mat_new <- merge_umi_mat_with_wt10(scmat = mat_new)
scdb_add_mat(id = "control_tetra_all_wt10", mat = mat_new)
```

#### Generate single-cell balanced kNN-graph
This step needs the file  
_data/tet_tko.bad_genes.txt_

```{r}
source("scripts/pipeline/gen_cgraph.R")
gen_cgraph(mat_nm = "control_tetra_all", force_recompute_cgraph = F)
gen_cgraph(mat_nm = "control_tetra_all_wt10", force_recompute_cgraph = F)
```

#### Transfer cell type annotation to chimera cells

```{r}
source("scripts/pipeline/transfer_cell_type_annotation.R")
transfer_color_chimera_tetraploid(mat_nm = "control_tetra_all_wt10")
```

#### Time tetraploid-complemented embryos based on their injected control cells

```{r}
source("scripts/pipeline/transfer_time_annotation.R")
source("scripts/pipeline/tetraploid_timing.R")
tetra_timing("control_tetra_all_wt10")
```

#### Export infered time and cell type into table

```{r}
source("scripts/pipeline/create_metadata.R")
create_control_tetraploid_metadata()
```

```{r}

```

<!--chapter:end:Control-tetraploid-complemented-embryos-processing.Rmd-->

---
jupyter:
  jupytext:
    formats: ipynb,Rmd
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.4
  kernelspec:
    display_name: R
    language: R
    name: ir353
---

```{r, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
setwd(here::here())
```

```{r}
library("metacell")
library("tgstat")
library("tgconfig")
library("Matrix")
library("dplyr")
```

```{r}
scdb_init("scrna_db/", force_reinit = T)
```

#### Remove extraembryonic ectoderm and parietal endoderm cells
#### Merge cells with the wildtype atlas

```{r}
source("scripts/pipeline/remove_exe_ectoderm_and_parietal_endo_cls.R")
source("scripts/pipeline/merge_umi_mat_with_wt10_umi_mat.R")

for (mat_nm in c("dko12_chim", "dko13_chim", "dko23_chim")) {
    print(mat_nm)
    mat <- scdb_mat(mat_nm)
    mat <- remove_exe_ectoderm_and_parietal_endoderm(mat_query = mat)
    mat_new <- merge_umi_mat_with_wt10(scmat = mat)
    scdb_add_mat(id = paste0(mat_nm, "_wt10"), mat = mat_new)
}
```

#### Generate single-cell balanced kNN-graph
This step needs the file  
_data/tet_tko.bad_genes.txt_

```{r}
source("scripts/pipeline/gen_cgraph.R")
gen_cgraph("dko12_chim_wt10", force_recompute_cgraph = F)
gen_cgraph("dko13_chim_wt10", force_recompute_cgraph = F)
gen_cgraph("dko23_chim_wt10", force_recompute_cgraph = F)
```

#### Transfer cell type annotation to TKO cells

```{r}
source("scripts/pipeline/transfer_cell_type_annotation.R")
transfer_color_chimera_tetraploid("dko12_chim_wt10")
transfer_color_chimera_tetraploid("dko13_chim_wt10")
transfer_color_chimera_tetraploid("dko23_chim_wt10")
```

#### Time chimera embryos based on their host and KO cells

```{r}
source("scripts/pipeline/transfer_time_annotation.R")
source("scripts/pipeline/dko_chim_timing.R")
dko_chim_timing("dko12_chim_wt10")
dko_chim_timing("dko13_chim_wt10")
dko_chim_timing("dko23_chim_wt10")
```

```{r}

```

<!--chapter:end:DKO-chimera-embryos-processing.Rmd-->

---
jupyter:
  jupytext:
    formats: ipynb,Rmd
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.4
  kernelspec:
    display_name: R
    language: R
    name: ir353
---

```{r, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
setwd(here::here())
```

```{r}
library("metacell")
library("tgstat")
library("Matrix")
library("dplyr")
```

```{r}
source("scripts/pipeline/proj_on_wt9_atlas.R")
source("scripts/pipeline/add_clone_information_to_scmat_metadata.R")
```

```{r}
scfigs_init("figs")
scdb_init("scrna_db/", force_reinit = TRUE)
```

#### Tet TKO chimera embryos

```{r}
source("scripts/gating/tko_chim_gating.R")
```

#### Tet TKO tetraploid embryos

```{r}
source("scripts/gating/tko_tetra_gating.R")
```

####  Control tetraploid embryos

```{r}
source("scripts/gating/control_tetra_gating.R")
```

####  Tet 1/2 DKO chimera embryos

```{r}
source("scripts/gating/dko12_chim_gating.R")
dko12_gating()
```

####  Tet 1/3 DKO chimera embryos

```{r}
source("scripts/gating/dko13_chim_gating.R")
dko13_gating()
```

####  Tet 2/3 DKO chimera embryos

```{r}
source("scripts/gating/dko23_chim_gating.R")
dko23_gating()
```

```{r}

```

<!--chapter:end:gating.Rmd-->

---
jupyter:
  jupytext:
    formats: ipynb,Rmd
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.4
  kernelspec:
    display_name: R
    language: R
    name: ir353
---

```{r, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
## The working directory should be the root of the cloned repository with scrna_db, figs, data as subfolders
setwd(here::here())
```

```{r}
library("dplyr")
library("zoo")
library("metacell")
library("tgstat")
library("ggpubr")
library("qvalue")
library("gridExtra")
library("Matrix")
library("tidyr")
library("ggplot2")
library("ggrepel")
```

```{r}
scdb_init("scrna_db/", force_reinit = T)
```

```{r}
source("scripts/pipeline/atlas_proj_wt10.R")
source("scripts/pipeline/transfer_time_annotation.R")
```

#### Generate Figure 1

```{r}
source("scripts/paper_figures/fig1.R")
generate_figure1_plots()
```

#### Generate Figure 2

```{r}
source("scripts/paper_figures/fig2.R")
generate_figure2_plots()
```

#### Generate Figure 3

```{r}
source("scripts/paper_figures/fig3.R")
generate_figure3_plots()
```

#### Generate Figure 4

```{r}
source("scripts/paper_figures/fig4.R")
generate_figure4_plots()
```

#### Generate Figure 6

```{r}
source("scripts/paper_figures/fig6.R")
generate_figure6_plots()
```

#### Generate Figure S2

```{r}
source("scripts/paper_figures/fig_s2.R")
generate_figure_s2_plots()
```

#### Generate Figure S3

```{r}
source("scripts/paper_figures/fig_s3.R")
generate_figure_s3_plots()
```

#### Generate Figure S4

```{r}
source("scripts/paper_figures/fig_s4.R")
generate_figure_s4_plots()
```

#### Generate Figure S5

```{r}
source("scripts/paper_figures/fig_s5.R")
generate_figure_s5_plots()
```

#### Generate Figure S6

```{r}
source("scripts/paper_figures/fig_s6.R")
generate_figure_s6_plots()
```

```{r}

```

<!--chapter:end:generate-figures.Rmd-->

---
jupyter:
  jupytext:
    formats: ipynb,Rmd
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.4
  kernelspec:
    display_name: R
    language: R
    name: ir353
---

```{r, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
setwd(here::here())
```

```{r}
library("metacell")
library("Matrix")
library("dplyr")
```

```{r}
scdb_init("scrna_db/", force_reinit = TRUE)
```

### Import Tet TKO chimera plates into metacell
This notebook requires that the UMI tables for each plate are available in the folder **data/umi.tables** . If they are not available, please run the cell below to download the necessary data

```{r}
# If you need to download the umi tables, please run the two lines below
if (0) {
    source("../scripts/download_data.R")
    download_umi_tables()
}
```

```{r}
source("scripts/pipeline/import_plates.R")
```

```{r}
mat_nm <- "tko_chim"

metadata <- read.csv(file = "data/plate_metadata.csv", stringsAsFactors = F)
genotype <- "TKO"
treatment <- c("Chimera assay")
metadata <- metadata[metadata$treatment %in% treatment, ]
metadata <- metadata[grep(pattern = genotype, x = metadata$genotype), ]

# remove additionally the plate 190730_P03. No embryo and facs data available. Only relevant for TKO Chimera
metadata <- metadata[metadata$plate != "190730_P03", ]

import_plates(mat_nm = mat_nm, metadata = metadata)
```

### Import Tet TKO tetraploid-complemented embryos

```{r}
mat_nm <- "tko_tetra"

metadata <- read.csv(file = "data/plate_metadata.csv", stringsAsFactors = F)
genotype <- "TKO"
treatment <- c("Tetraploid complementation assay")
metadata <- metadata[metadata$treatment %in% treatment, ]
metadata <- metadata[grep(pattern = genotype, x = metadata$genotype), ]

import_plates(mat_nm = mat_nm, metadata = metadata)
```

```{r}
# There are four mixed plates containing both tetraploid-complemented control and Tet TKO embryos
# We additionally ignore the cells from control embryos
mat <- scdb_mat("tko_tetra")

ctrl_cells <- mat@cell_metadata$cell[grep(pattern = "Ctrl", x = mat@cell_metadata$embryo)]

mat <- scm_ignore_cells(scmat = mat, ig_cells = union(mat@ignore_cells, ctrl_cells))

scdb_add_mat(id = "tko_tetra", mat = mat)
```

### Import Control tetraploid-complemented embryos

```{r}
mat_nm <- "control_tetra"

metadata <- read.csv(file = "data/plate_metadata.csv", stringsAsFactors = F)
genotype <- "Control"
treatment <- c("Tetraploid complementation assay")
metadata <- metadata[metadata$treatment %in% treatment, ]
metadata <- metadata[grep(pattern = genotype, x = metadata$genotype), ]

import_plates(mat_nm = mat_nm, metadata = metadata)
```

```{r}
# There are four mixed plates containing both tetraploid-complemented control and Tet TKO embryos
# We additionally ignore the cells from TKO embryos
mat <- scdb_mat("control_tetra")

tko_cells <- mat@cell_metadata$cell[grep(pattern = "TKO", x = mat@cell_metadata$embryo)]
length(tko_cells)

mat <- scm_ignore_cells(scmat = mat, ig_cells = union(mat@ignore_cells, tko_cells))

scdb_add_mat(id = "control_tetra", mat = mat)
```

### Import Tet 1/2 DKO chimera embryos

```{r}
mat_nm <- "dko12_chim"

metadata <- read.csv(file = "data/plate_metadata.csv", stringsAsFactors = F)
genotype <- "Tet1/2 DKO"
treatment <- c("Chimera assay")
metadata <- metadata[metadata$treatment %in% treatment, ]
metadata <- metadata[grep(pattern = genotype, x = metadata$genotype), ]

import_plates(mat_nm = mat_nm, metadata = metadata)
```

### Import Tet 1/3 DKO chimera embryos

```{r}
mat_nm <- "dko13_chim"

metadata <- read.csv(file = "data/plate_metadata.csv", stringsAsFactors = F)
genotype <- "Tet1/3 DKO"
treatment <- c("Chimera assay")
metadata <- metadata[metadata$treatment %in% treatment, ]
metadata <- metadata[grep(pattern = genotype, x = metadata$genotype), ]

import_plates(mat_nm = mat_nm, metadata = metadata)
```

### Import Tet 2/3 DKO chimera embryos

```{r}
mat_nm <- "dko23_chim"

metadata <- read.csv(file = "data/plate_metadata.csv", stringsAsFactors = F)
genotype <- "Tet2/3 DKO"
treatment <- c("Chimera assay")
metadata <- metadata[metadata$treatment %in% treatment, ]
metadata <- metadata[grep(pattern = genotype, x = metadata$genotype), ]

import_plates(mat_nm = mat_nm, metadata = metadata)
```

### Import AAV-Cre delivered whole-embryo TKOs

```{r}
mat_nm <- "tko_germline"

metadata <- read.csv(file = "data/plate_metadata.csv", stringsAsFactors = F)
genotype <- "TKO"
treatment <- c("AAV Cre-delivered Tet1/2/3 triple KO")
metadata <- metadata[metadata$treatment %in% treatment, ]
metadata <- metadata[grep(pattern = genotype, x = metadata$genotype), ]

import_plates(mat_nm = mat_nm, metadata = metadata)
```

```{r}
#### next step: gating of populations
```

```{r}

```

<!--chapter:end:import_plates.Rmd-->

---
jupyter:
  jupytext:
    formats: ipynb,R:light,Rmd
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.4
  kernelspec:
    display_name: R
    language: R
    name: ir
---

# Methylation analysis


## Initialization 

```{r}
if (!require("tet.gastru")) {
    if (!require("devtools")) {
        install.packages("devtools")
    }
    devtools::install(here("scripts/tet.gastru"))
}
library(tet.gastru)
init()
```

## Define genomic contexts


#### CTCF motif

```{r}
gvtrack.create("ctcf", "ctcf_motif", "global.percentile.max")
```

#### CpG islands

```{r}
cgi_intervs <- gintervals.load("intervs.global.cpgIsland")
```

#### Time of replication (TOR)

```{r}
gvtrack.create("tor", "Encode.esd3.replichip.rep2", "avg")
gvtrack.iterator("tor", sshift = -15000, eshift = 15000)
```

#### K4me3/K27me3 peaks


Due to the large size of the ENCODE tracks which generated the CHIP peaks, we supply here also the result of our computation. You can see the details at: `scripts/tet.gastru/R/genomic-context.R`

* `define_k4me3_peaks()`
* `define_k27me_peaks()`
* `define_atac_peaks()`

In a nutshell - we aggregate functional annotation using data from ENCODE (Bing Ren's lab), and from the following paper: 

Xiang, Y., Zhang, Y., Xu, Q. et al. Epigenomic analysis of gastrulation identifies a unique chromatin state for primed pluripotency. Nat Genet 52, 95–105 (2020). https://doi.org/10.1038/s41588-019-0545-1


```{r}
k4me3_peaks <- fread(here("data/k4me3_peaks.tsv"))
k27me3_peaks <- fread(here("data/k27me3_peaks.tsv"))
atac_peaks <- fread(here("data/atac_peaks.tsv"))
```

#### Local 'dips' in methylation 


A local reduction in methylation can be an indicator for a functional element (e.g. enhancer). We screen for such regions at methylation of e8.5:

```{r}
meth_dips <- meth_dip_screen("e8_5", dip_size = 300, flank_size = 500, margin = 50, min_diff = 0.4, min_cov = 50, ret_values = TRUE) %cache_df% here("output/e8.5_dips.tsv") %>% as_tibble()
```

## Extract methylation data 


We will start by extracting all methylation data:


### All CpGs


We extract methylation in smoothed over 200bp from each side of the CpG, and not smoothed:

```{r}
meth_df_comp <- gextract_meth(
    c(
        comb_tracks,
        "e8_5",
        "Zhang_Nature_Genetics_2017.Ect_mCG",
        "Zhang_Nature_Genetics_2017.End_mCG",
        "Zhang_Nature_Genetics_2017.Mes_mCG",
        "Zhang_Nature_Genetics_2017.E65Epi_mCG"
    ),
    names = c(comb_names, "e8.5", "ecto", "endo", "meso", "epi"),
    extract_meth_calls = TRUE,
    d_expand = 200,
    annot_tracks = "ctcf"
) %>%
    mutate(e7.5.meth = ecto.meth + endo.meth + meso.meth, e7.5.cov = ecto.cov + endo.cov + meso.cov, e7.5 = e7.5.meth / e7.5.cov) %cache_df%
    here("output/cpg_meth_expand_200_comp.tsv") %>%
    as_tibble()
nrow(meth_df_comp)
head(meth_df_comp)
```

```{r}
meth_df_comp_non_smoothed <- gextract_meth(
    c(
        comb_tracks,
        "e8_5",
        "Zhang_Nature_Genetics_2017.Ect_mCG",
        "Zhang_Nature_Genetics_2017.End_mCG",
        "Zhang_Nature_Genetics_2017.Mes_mCG",
        "Zhang_Nature_Genetics_2017.E65Epi_mCG"
    ),
    names = c(comb_names, "e8.5", "ecto", "endo", "meso", "epi"),
    extract_meth_calls = TRUE,
    d_expand = 200,
    annot_tracks = "ctcf"
) %>%
    mutate(e7.5.meth = ecto.meth + endo.meth + meso.meth, e7.5.cov = ecto.cov + endo.cov + meso.cov, e7.5 = e7.5.meth / e7.5.cov) %cache_df%
    here("output/cpg_meth_comp.tsv") %>%
    as_tibble()
nrow(meth_df_comp_non_smoothed)
head(meth_df_comp_non_smoothed)
```

## Distribution of TET-TKO vs WT methylation (Figure 5B)

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
min_cov <- 20
plot_smooth_cpg_scatter(meth_df_comp, "ctrl", "tet_tko", min_cov, "Control", "Tet-TKO")
```

```{r}
options(repr.plot.width = 5, repr.plot.height = 6)
plot_smooth_legend(c("white", "lightblue", "blue", "darkblue", "yellow", "gold", "orange", "red", "darkred"))
```

## Methylation in TADs


We took TAD defnitions from Nagano et al., 2017:

```{r}
tad_intervs <- gintervals.load("scell.nextera.pool_good_hyb_es_b_TADs") %>% as_tibble()
dim(tad_intervs)
```

We can then extract average methylation in TADs and then remove from it ('punctuate') the methylation calls that are at k4me3 or k27me3 peaks, or that are at methylation dips or ATAC peaks:

```{r}
tads_meth <- get_cpg_meth_tads(tad_intervs, meth_dips, k4me3_peaks, k27me3_peaks, atac_peaks) %cache_df% here("output/tads_meth.tsv") %>% as_tibble()
```

### Compare TAD methylation in Ctrl and TET-tko (Figure 5C)

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
df <- tads_meth %>%
    filter(ctrl.cov >= 3e3, tet_tko.cov >= 3e3) %>%
    mutate(l = end - start) %>%
    filter(l >= 5e4) %>%
    mutate(ctrl_bins = cut(ctrl, c(0.7, 0.8, 0.83, 0.86, 0.89, 1))) %>%
    select(chrom, start, end, ctrl_bins, Ctrl = ctrl, TKO = tet_tko) %>%
    gather("type", "meth", -(chrom:ctrl_bins))

df %>% count(ctrl_bins)

p_tads_boxp <- df %>%
    ggplot(aes(x = ctrl_bins, y = meth, fill = type)) +
    geom_boxplot(lwd = 0.01, outlier.size = 0.001, fatten = 0.1) +
    scale_fill_manual(name = "", values = c("Ctrl" = "#00468BFF", "TKO" = "#ED0000FF")) +
    theme(aspect.ratio = 1) +
    vertical_labs() +
    xlab("Ctrl meth.") +
    ylab("TADs meth.")
p_tads_boxp
```

```{r}
df <- tads_meth %>%
    filter(ctrl.cov >= 3e3, tet_tko.cov >= 3e3) %>%
    mutate(l = end - start) %>%
    filter(l >= 5e4) %>%
    summarise(
        tet_tko_mean = mean(tet_tko),
        tet_tko_sd = sd(tet_tko, na.rm = TRUE),
        ctrl_mean = mean(ctrl),
        ctrl_sd = sd(ctrl, na.rm = TRUE),
        e7.5_mean = mean(e7.5),
        e7.5_sd = sd(e7.5, na.rm = TRUE),
        e8.5_mean = mean(e8.5),
        e8.5_sd = sd(e8.5, na.rm = TRUE)
    )
df
df %>%
    summarise_at(vars(everything()), function(x) round(x, digits = 2))

ks.test(tads_meth$ctrl, tads_meth$tet_tko)
```

We can see that that the variation in background methylation between TADs was greatly diminished in Tet-TKO cells.


### Figure S6A

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
df <- tads_meth %>%
    filter(ctrl.cov >= 3e3, tet_tko.cov >= 3e3) %>%
    mutate(l = end - start) %>%
    filter(l >= 5e4) %>%
    mutate(ctrl_bins = cut(ctrl, c(0.7, 0.8, 0.83, 0.86, 0.89, 1))) %>%
    select(chrom, start, end, ctrl_bins, Ctrl = ctrl, WT = e8.5) %>%
    gather("type", "meth", -(chrom:ctrl_bins))

df %>% count(ctrl_bins)

p_tads_wt_boxp <- df %>%
    ggplot(aes(x = ctrl_bins, y = meth, fill = type)) +
    geom_boxplot(lwd = 0.01, outlier.size = 0.001, fatten = 0.1) +
    scale_fill_manual(name = "", values = c("Ctrl" = "#00468BFF", "WT" = "#42B540FF")) +
    theme(aspect.ratio = 1) +
    vertical_labs() +
    xlab("Ctrl meth.") +
    ylab("TADs meth.")
p_tads_wt_boxp
```

We can see that our analysis infers higher methylation levels in the embryo compared to some previously published WT data. This is likely due to the elimination of extraembryonic ectoderm from the analysis, which represents a more hypomethylated (Smith et al., 2017) cell population that can affect the estimation of average methylation when not excluded.


## TAD methylation at different TOR regimes (Figure 5D)

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
p_tads_tor <- tads_meth %>%
    filter(ctrl.cov >= 3e3, tet_tko.cov >= 3e3) %>%
    mutate(l = end - start) %>%
    filter(l >= 5e4) %>%
    rename(Ctrl = ctrl, TKO = tet_tko) %>%
    pivot_longer(Ctrl:TKO, names_to = "type", values_to = "meth") %>%
    ggplot(aes(x = -tor, y = meth, color = type)) +
    geom_smooth(method = lm, se = FALSE) +
    geom_point(size = 0.01, alpha = 0.1) +
    theme(aspect.ratio = 1) +
    xlab("TOR") +
    ylab("Methylation") +
    scale_color_manual(name = "", values = c("Ctrl" = "#00468BFF", "TKO" = "#ED0000FF")) +
    scale_x_continuous(breaks = c(-1, 1), labels = c("Early", "Late")) +
    vertical_labs()
p_tads_tor

tads_meth %>%
    filter(ctrl.cov >= 3e3, tet_tko.cov >= 3e3) %>%
    mutate(l = end - start) %>%
    filter(l >= 5e4) %>%
    pivot_longer(ctrl:tet_tko, names_to = "type", values_to = "meth") %>%
    group_by(type) %>%
    summarise(cor = cor(meth, tor))
```

We see that that lower background methylation is linked with early replicating TADs in control and a reciprocal effect in Tet-TKO.
These data support a role for widespread TET-mediated de-methylation in early replicating TADs, which is lost upon Tet knockout. Such de-methylation may rely on the enhanced accessibility of early replicating domains as part of the chromosomal A-compartment (López-Moyado et al., 2019; Pope et al., 2014).


## Low methylation hotspots


We extract chromatin marks at hotspots of low methylation (control methylation <0.25):

```{r}
define_chip_vtracks()

low_meth_df <- meth_df_comp %>%
    select(chrom, start, end, ctrl, tet_tko, e8.5, ctrl.cov, tet_tko.cov, e8.5.cov) %>%
    gextract.left_join(
        c(
            "pmax(-log2(1-ecto_k27me3),-log2(1-endo_k27me3),-log2(1-meso_k27me3),-log2(1-ps_k27me3),-log2(1-e6_epi_k27me3),-log2(1-e5_epi_k27me3),-log2(1-e6_ve_k27me3),-log2(1-e10_fc_k27me3),-log2(1-e10_forebrain_k27me3),-log2(1-e10_heart_k27me3),-log2(1-e10_hindbrain_k27me3),-log2(1-e10_limb_k27me3),-log2(1-e10_midbrain_k27me3),-log2(1-es_k27me3), na.rm=TRUE)",
            "pmax(-log2(1-e10_limb_k4me3),-log2(1-e10_heart_k4me3),-log2(1-e10_midbrain_k4me3),-log2(1-e10_hindbrain_k4me3),-log2(1-e10_forebrain_k4me3),-log2(1-e11_liver_k4me3),-log2(1-ecto_k4me3),-log2(1-endo_k4me3),-log2(1-meso_k4me3),-log2(1-ps_k4me3),-log2(1-e6_epi_k4me3),-log2(1-e5_epi_k4me3),-log2(1-e6_ve_k4me3), na.rm=TRUE)"
        ),
        iterator = ., intervals = ., colnames = c("k27me3", "k4me3")
    ) %>%
    select(-(chrom1:end1)) %cache_df%
    here("output/meth_marks_smoothed.tsv") %>%
    as_tibble()
```

### Figure S6B

```{r}
options(repr.plot.width = 14, repr.plot.height = 7)
df <- low_meth_df %>%
    filter(ctrl.cov >= 10) %>%
    mutate(low_meth = ifelse(ctrl <= 0.25, "Low meth.", "Other"))

p_k4me3 <- df %>%
    ggplot(aes(x = k4me3, color = low_meth, y = 1 - ..y..)) +
    stat_ecdf() +
    scale_color_manual("", values = c("darkblue", "darkgray")) +
    scale_y_continuous(labels = scales::percent) +
    ylab("% of loci") +
    xlab("H3K4me3")

p_k27me3 <- df %>%
    ggplot(aes(x = k27me3, color = low_meth, y = 1 - ..y..)) +
    stat_ecdf() +
    scale_color_manual("", values = c("darkblue", "darkgray")) +
    scale_y_continuous(labels = scales::percent) +
    ylab("% of loci") +
    xlab("H3K27me3")

p_k4me3 + p_k27me3
```

## K4me3/K27me3 hospots (Figure 5E)


### K4me3 hotspots

```{r}
d_k4me3 <- extract_meth_at_context(k4me3_peaks) %cache_df% here("output/k4me3_meth_comp.tsv") %>% as_tibble()
```

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
df_k4me3 <- d_k4me3 %>%
    filter(ctrl.cov >= 50, tet_tko.cov >= 50) %>%
    gintervals.neighbors1(k27me3_peaks) %>%
    mutate(type = ifelse(dist == 0, "bivalent", "other")) %>%
    select(-(chrom1:dist)) %>%
    mutate(l = end - start) %>%
    filter(abs(l) >= 3e3)

p_k4me3 <- df_k4me3 %>%
    ggplot(aes(x = ctrl, y = tet_tko, color = type)) +
    geom_point(size = 0.1, alpha = 0.5) +
    theme(aspect.ratio = 1) +
    geom_abline() +
    theme(aspect.ratio = 1) +
    scale_color_manual(name = "", values = c("bivalent" = "darkred", "other" = "darkgray")) +
    ggtitle("K4me3") +
    xlab("WT") +
    ylab("TKO ") +
    xlim(0, 1) +
    ylim(0, 1)

p_k4me3
```

```{r}
df_k4me3 %>% count(type)
nrow(df_k4me3)
```

```{r}
df_k4me3 %>%
    select(chrom:end, ctrl, tet_tko, type) %>%
    mutate(diff = tet_tko - ctrl) %>%
    filter(diff >= 0) %>%
    summarise(min_diff = min(diff), max_diff = max(diff), m = mean(diff), sd = sd(diff)) %>%
    summarise_at(vars(everything()), scales::percent)
```

### K27me3 hotspots

```{r}
d_k27me3 <- extract_meth_at_context(k27me3_peaks) %cache_df% here("output/k27me3_meth_comp.tsv") %>% as_tibble()
```

Polycomb without methylation of k4me3 regions ("punctuated"):

```{r tags=c()}
d_k27me3_punc <- gextract_meth(c(comb_tracks, "e8_5", "Zhang_Nature_Genetics_2017.Ect_mCG", "Zhang_Nature_Genetics_2017.End_mCG", "Zhang_Nature_Genetics_2017.Mes_mCG", "Zhang_Nature_Genetics_2017.E65Epi_mCG"), names = c(comb_names, "e8.5", "ecto", "endo", "meso", "epi"), extract_meth_calls = TRUE, intervals = k27me3_peaks, join_intervals = TRUE) %>%
    mutate(e7.5.meth = ecto.meth + endo.meth + meso.meth, e7.5.cov = ecto.cov + endo.cov + meso.cov, e7.5 = e7.5.meth / e7.5.cov) %>%
    gintervals.neighbors1(k4me3_peaks) %>%
    filter(dist != 0) %>%
    group_by(chrom1, start1, end1) %>%
    summarise_at(vars(ends_with("cov"), ends_with("meth")), sum, na.rm = TRUE) %>%
    rename(chrom = chrom1, start = start1, end = end1) %>%
    mutate(ctrl = ctrl.meth / ctrl.cov, tet_tko = tet_tko.meth / tet_tko.cov, e8.5 = e8.5.meth / e8.5.cov, e7.5 = e7.5.meth / e7.5.cov, epi = epi.meth / epi.cov) %>%
    ungroup() %cache_df% here("output/k27me3_meth_comp_k4me3_punc.tsv") %>%
    as_tibble()
```

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
df <- d_k27me3 %>%
    filter(ctrl.cov >= 50, tet_tko.cov >= 50) %>%
    gintervals.neighbors1(k4me3_peaks) %>%
    mutate(type = ifelse(dist == 0, "bivalent", "other")) %>%
    select(-(chrom1:dist)) %>%
    mutate(l = end - start) %>%
    filter(abs(l) >= 3e3)
p_k27me3 <- df %>%
    ggplot(aes(x = ctrl, y = tet_tko, color = type)) +
    geom_point(size = 0.5) +
    theme(aspect.ratio = 1) +
    geom_abline() +
    theme_bw() +
    theme(aspect.ratio = 1) +
    scale_color_manual(name = "", values = c("bivalent" = "darkred", "other" = "darkgray")) +
    ggtitle("H3K27me3") +
    xlab("Ctrl") +
    ylab("TKO")

p_k27me3
```

```{r}
df %>% count(type)
nrow(df)
```

H3k27me3 hotspots while removing H3K4me3 methylation:

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
df_punc <- d_k27me3_punc %>%
    filter(ctrl.cov >= 50, tet_tko.cov >= 50) %>%
    gintervals.neighbors1(k4me3_peaks) %>%
    mutate(type = ifelse(dist == 0, "bivalent", "other")) %>%
    select(-(chrom1:dist)) %>%
    mutate(l = end - start) %>%
    filter(abs(l) >= 3e3)

p_k27_punc <- df_punc %>%
    ggplot(aes(x = ctrl, y = tet_tko, color = type)) +
    geom_point(size = 0.1, alpha = 0.5) +
    theme(aspect.ratio = 1) +
    geom_abline() +
    theme(aspect.ratio = 1) +
    scale_color_manual(name = "", values = c("bivalent" = "darkred", "other" = "darkgray")) +
    ggtitle("H3K27me3") +
    xlab("Ctrl") +
    ylab("TKO")
p_k27_punc
```

```{r}
df_punc %>%
    select(chrom:end, ctrl, tet_tko) %>%
    mutate(diff = tet_tko - ctrl) %>%
    filter(diff >= 0) %>%
    summarise(min_diff = min(diff), max_diff = max(diff), m = mean(diff), sd = sd(diff)) %>%
    summarise_at(vars(everything()), scales::percent)
```

```{r}
bind_rows(df_k4me3, df_punc) %>%
    select(chrom:end, ctrl, tet_tko, type) %>%
    mutate(diff = tet_tko - ctrl) %>%
    filter(diff >= 0) %>%
    summarise(min_diff = min(diff), max_diff = max(diff), m = mean(diff), sd = sd(diff)) %>%
    summarise_at(vars(everything()), scales::percent)
```

```{r}
n_above <- bind_rows(df_k4me3, df_punc) %>%
    select(chrom:end, ctrl, tet_tko, type) %>%
    filter(ctrl <= 0.25, tet_tko >= 0.5) %>%
    nrow()

scales::percent(n_above / nrow(bind_rows(df_k4me3, df_punc)))
```

```{r}
df_punc %>% count(type)
nrow(df_punc)
```

## Loci linked with CTCF occupancy (Figure 5F)

```{r}
tet_df <- meth_df_comp %>% filter(ctcf >= 0.999)
tet_df <- tet_df %>%
    gintervals.neighbors1(k4me3_peaks) %>%
    mutate(type = ifelse(dist == 0, "k4me3", "other")) %>%
    select(chrom:end, ctrl.cov, tet_tko.cov, ctrl, tet_tko, ctcf, type)
```

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
min_cov <- 20
p_ctcf <- tet_df %>%
    filter(ctrl.cov >= min_cov, tet_tko.cov >= min_cov) %>%
    ggplot(aes(x = ctrl, y = tet_tko)) +
    geom_point(size = 0.001, alpha = 0.5) +
    theme(aspect.ratio = 1) +
    geom_abline(color = "red") +
    facet_wrap(~type) +
    ggtitle("CTCF peaks") +
    xlab("WT") +
    ylab("TKO")

p_ctcf
```

```{r}
tet_df %>%
    filter(ctrl.cov >= min_cov, tet_tko.cov >= min_cov) %>%
    count(type)
```

## Methylation at Exons and Non-H3K4me3 Promoters (Figure 5G)


### Exons

```{r}
d_exons <- extract_meth_at_context("intervs.global.exons") %cache_df% here("output/exons_meth_comp.tsv") %>% as_tibble()
d_exons <- d_exons %>%
    gintervals.neighbors1(k4me3_peaks) %>%
    filter(abs(dist) > 0) %>%
    select(-(chrom1:dist)) %>%
    gintervals.neighbors1(k27me3_peaks) %>%
    filter(abs(dist) > 0) %>%
    select(-(chrom1:dist))
```

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
p_exons <- d_exons %>%
    filter(ctrl.cov >= 30, tet_tko.cov >= 30) %>%
    ggplot(aes(x = ctrl, y = tet_tko)) +
    geom_point(size = 0.1, alpha = 0.5) +
    theme(aspect.ratio = 1) +
    geom_abline() +
    theme(aspect.ratio = 1) +
    ggtitle("Exons") +
    xlab("Ctrl") +
    ylab("TKO")
p_exons
```

```{r}
d_exons %>%
    filter(ctrl.cov >= 30, tet_tko.cov >= 30) %>%
    nrow()
```

### Compare with the WT (Figure S6C):

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
p_exons_vs_wt <- d_exons %>%
    filter(e7.5.cov >= 30, tet_tko.cov >= 30) %>%
    ggplot(aes(x = e7.5, y = tet_tko)) +
    geom_point(size = 0.1, alpha = 0.5) +
    theme(aspect.ratio = 1) +
    geom_abline() +
    theme(aspect.ratio = 1) +
    ggtitle("Exons") +
    xlab("WT") +
    ylab("TKO")
p_exons_vs_wt
```

### Promoters

```{r}
d_promoters <- extract_meth_at_context(get_promoters()) %cache_df% here("output/promoter_meth_comp.tsv") %>% as_tibble()
```

Annotate promoters to 'k4me3', 'k27me3', 'bivalent' and other ('none'):

```{r}
d_promoters <- d_promoters %>%
    gintervals.neighbors1(k4me3_peaks) %>%
    mutate(k4me3 = dist == 0) %>%
    select(-(chrom1:dist)) %>%
    gintervals.neighbors1(k27me3_peaks) %>%
    mutate(k27me3 = dist == 0) %>%
    select(-(chrom1:dist)) %>%
    mutate(type = case_when(k27me3 & k4me3 ~ "bivalent", k4me3 ~ "k4me3", k27me3 ~ "k27me3", TRUE ~ "none"))
```

Add time of replication annotation:

```{r}
d_promoters <- gextract.left_join("tor", intervals = d_promoters, iterator = d_promoters) %>%
    as_tibble() %>%
    select(-(chrom1:end1))
```

We match the promoters with marginal expression data from the gastrulation atlas (Mittnenzweig, Meishar et. al. Cell 2021):

```{r}
atlas_egc <- readr::read_rds(here("data/atlas_egc.rds"))
atlas_marginal <- tibble(gene = rownames(atlas_egc), expr = log2(rowSums(atlas_egc) + 1e-5))
```

#### Non-H3Kme3 promoters


We require a coverage of at least 30: 

```{r}
df <- d_promoters %>%
    filter(type == "none", ctrl.cov >= 30, tet_tko.cov >= 30) %>%
    left_join(get_promoters() %>% select(chrom:end, gene = geneSymbol)) %>%
    left_join(atlas_marginal %>% separate_rows(gene, sep = ";")) %>%
    gintervals.neighbors1(cgi_intervs) %>%
    mutate(cgi = ifelse(dist == 0, "cgi", "other")) %>%
    filter(!is.na(expr))
```

```{r}
df %>%
    filter(!is.na(ctrl), !is.na(tet_tko)) %>%
    nrow()
```

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
p_prom_none <- df %>%
    ggplot(aes(x = ctrl, y = tet_tko)) +
    geom_point(size = 0.1, alpha = 0.5) +
    theme(aspect.ratio = 1) +
    geom_abline() +
    theme(aspect.ratio = 1) +
    ggtitle("Non-H3K4me3 Promoters") +
    xlab("Ctrl") +
    ylab("TKO")
p_prom_none
```

### Compare with the WT (Figure S6C):

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
p_prom_none_vs_wt <- df %>%
    filter(e7.5.cov >= 30, tet_tko.cov >= 30) %>%
    ggplot(aes(x = e7.5, y = tet_tko)) +
    geom_point(size = 0.1, alpha = 0.5) +
    theme(aspect.ratio = 1) +
    geom_abline() +
    theme(aspect.ratio = 1) +
    ggtitle("Non-H3K4me3 Promoters") +
    xlab("WT") +
    ylab("TKO")
p_prom_none_vs_wt
```

## Promoter methylation vs expression (Figure S6D/S6E)


Load expression data of all the types and of epiblast:

```{r}
egc_bulk <- fread(here("data/eg_bulk.tsv")) %>% as_tibble()
egc_epiblast <- fread(here("data/eg_epiblast.tsv")) %>% as_tibble()
```

We generate a table for all the covered promoters in wt.

```{r}
promoters_table <- d_promoters %>%
    left_join(get_promoters() %>% select(chrom:end, strand, geneSymbol), by = c("chrom", "start", "end")) %>%
    filter(ctrl.cov >= 30) %>%
    separate_rows(geneSymbol, sep = ";") %>%
    # when there are alternative promoters - keep the one with lower methylation
    arrange(geneSymbol, ctrl) %>%
    distinct(geneSymbol, .keep_all = TRUE) %>%
    # add expression data
    left_join(egc_bulk %>% separate_rows(gene, sep = ";") %>% rename(geneSymbol = gene), by = "geneSymbol") %>%
    mutate_at(vars(ends_with("bulk")), function(x) log2(x + 1e-5)) %cache_df%
    here("output/promoter_table.tsv") %>%
    as_tibble()
```

```{r}
promoters_table %>%
    filter(wt_bulk >= -14) %>%
    count(type) %>%
    mutate(p = n / sum(n), perc = scales::percent(p))
```

```{r}
p_expr_marks <- promoters_table %>%
    mutate(expr = cut(wt_bulk, c(-17, -16, -14, 0), labels = c("low", "mid", "high"))) %>%
    filter(!is.na(expr)) %>%
    count(expr, type) %>%
    ggplot(aes(x = expr, y = n, fill = type)) +
    geom_col() +
    scale_fill_manual(name = "", values = c("bivalent" = "darkred", "k27me3" = "darkblue", "k4me3" = "darkgreen", "none" = "darkgray")) +
    ylab("# of promoters") +
    xlab("Gene expression")

p_expr_marks
```

```{r}
define_chip_vtracks()
intervs <- promoters_table %>% select(chrom:end, geneSymbol, wt_bulk)
marks_expr <- gextract.left_join(
    c(
        "pmax(-log2(1-ecto_k27me3),-log2(1-endo_k27me3),-log2(1-meso_k27me3),-log2(1-ps_k27me3),-log2(1-e6_epi_k27me3),-log2(1-e5_epi_k27me3),-log2(1-e6_ve_k27me3),-log2(1-e10_fc_k27me3),-log2(1-e10_forebrain_k27me3),-log2(1-e10_heart_k27me3),-log2(1-e10_hindbrain_k27me3),-log2(1-e10_limb_k27me3),-log2(1-e10_midbrain_k27me3),-log2(1-es_k27me3), na.rm=TRUE)",
        "pmax(-log2(1-e10_limb_k4me3),-log2(1-e10_heart_k4me3),-log2(1-e10_midbrain_k4me3),-log2(1-e10_hindbrain_k4me3),-log2(1-e10_forebrain_k4me3),-log2(1-e11_liver_k4me3),-log2(1-ecto_k4me3),-log2(1-endo_k4me3),-log2(1-meso_k4me3),-log2(1-ps_k4me3),-log2(1-e6_epi_k4me3),-log2(1-e5_epi_k4me3),-log2(1-e6_ve_k4me3), na.rm=TRUE)"
    ),
    iterator = intervs, intervals = intervs, colnames = c("k27me3", "k4me3")
) %>%
    select(-(chrom1:end1)) %cache_df% here("output/marks_expr.tsv") %>%
    as_tibble()
```

```{r}
options(repr.plot.width = 14, repr.plot.height = 5)
marks_expr %>%
    mutate(expr = cut(wt_bulk, c(-17, -16, -14, 0), labels = c("low", "mid", "high"))) %>%
    filter(!is.na(expr)) %>%
    ggplot(aes(x = k4me3, y = k27me3)) +
    geom_point(size = 0.1) +
    facet_wrap(~expr) +
    theme(aspect.ratio = 1)
```


```{r}
marks_expr %>%
    mutate(expr = cut(wt_bulk, c(-17, -16, -14, 0), labels = c("low", "mid", "high"))) %>%
    filter(!is.na(k4me3)) %>%
    count(expr)
```

### Figure S6D

```{r}
options(repr.plot.width = 10, repr.plot.height = 4)
p_k4me3 <- marks_expr %>%
    mutate(expr = cut(wt_bulk, c(-17, -16, -14, 0), labels = c("low", "mid", "high"))) %>%
    filter(!is.na(expr)) %>%
    ggplot(aes(x = k4me3, color = expr, y = 1 - ..y..)) +
    geom_vline(xintercept = 7, linetype = "dashed") +
    stat_ecdf() +
    ggsci::scale_color_npg(name = "Expression") +
    ylab("% of promoters") +
    scale_y_continuous(label = scales::percent)
p_k27me3 <- marks_expr %>%
    mutate(expr = cut(wt_bulk, c(-17, -16, -14, 0), labels = c("low", "mid", "high"))) %>%
    filter(!is.na(expr)) %>%
    ggplot(aes(x = k27me3, color = expr, y = 1 - ..y..)) +
    geom_vline(xintercept = 7, linetype = "dashed") +
    stat_ecdf() +
    ggsci::scale_color_npg(name = "Expression") +
    ylab("% of promoters") +
    scale_y_continuous(label = scales::percent)

p_k4me3 + p_k27me3
```

```{r}
promoters_table %>%
    filter(type == "none") %>%
    mutate(expr = cut(wt_bulk, c(-17, -16, -14, 0), labels = c("Low Expr.", "Mid Expr.", "High Expr."))) %>%
    filter(!is.na(expr)) %>%
    filter(ctrl.cov >= 30, tet_tko.cov >= 30) %>%
    count(expr)
```

### Figure S6E

```{r}
options(repr.plot.width = 14, repr.plot.height = 4)
p_wt_tko_expr <- promoters_table %>%
    filter(type == "none") %>%
    mutate(expr = cut(wt_bulk, c(-17, -16, -14, 0), labels = c("Low Expr.", "Mid Expr.", "High Expr."))) %>%
    filter(!is.na(expr)) %>%
    filter(ctrl.cov >= 30, tet_tko.cov >= 30) %>%
    ggplot(aes(x = ctrl, y = tet_tko)) +
    geom_abline(linetype = "dashed") +
    geom_point(size = 0.1) +
    facet_wrap(~expr) +
    theme(aspect.ratio = 1) +
    vertical_labs() +
    xlab("Ctrl.") +
    ylab("TKO")

p_wt_tko_expr
```

## Putative enhancers


We extract CHIP data for methylation dips which do not have a known functional annotation (h4me3, k27me3 or promoters):

```{r}
bg_intervs <- gintervals.diff(meth_dips, k4me3_peaks) %>%
    gintervals.diff(k27me3_peaks) %>%
    gintervals.diff(get_promoters())
```

```{r}
define_chip_vtracks()

meth_dips_chip <- bg_intervs %>%
    gextract.left_join(
        c(
            "pmax(-log2(1-e10_limb_k4me1),-log2(1-e10_heart_k4me1),-log2(1-e10_fc_k4me1),-log2(1-e10_midbrain_k4me1),-log2(1-e10_hindbrain_k4me1),-log2(1-e10_forebrain_k4me1), na.rm=TRUE)",
            "pmax(-log2(1-e10_limb_k27ac),-log2(1-e10_heart_k27ac),-log2(1-e10_midbrain_k27ac),-log2(1-e10_hindbrain_k27ac),-log2(1-e10_forebrain_k27ac),-log2(1-ecto_k27ac),-log2(1-endo_k27ac),-log2(1-meso_k27ac),-log2(1-ps_k27ac),-log2(1-e6_epi_k27ac),-log2(1-e6_ve_k27ac), na.rm=TRUE)"
        ),
        iterator = ., intervals = ., colnames = c("k4me1", "k27ac")
    ) %>%
    select(-(chrom1:end1)) %cache_df% here("output/meth_dips_chip.tsv") %>%
    as_tibble()
```

We extract the same data for all all other non-functionaly annotated regions:

```{r}
chip_bg <- gextract(
    c(
        "pmax(-log2(1-e10_limb_k4me1),-log2(1-e10_heart_k4me1),-log2(1-e10_fc_k4me1),-log2(1-e10_midbrain_k4me1),-log2(1-e10_hindbrain_k4me1),-log2(1-e10_forebrain_k4me1), na.rm=TRUE)",
        "pmax(-log2(1-e10_limb_k27ac),-log2(1-e10_heart_k27ac),-log2(1-e10_midbrain_k27ac),-log2(1-e10_hindbrain_k27ac),-log2(1-e10_forebrain_k27ac),-log2(1-ecto_k27ac),-log2(1-endo_k27ac),-log2(1-meso_k27ac),-log2(1-ps_k27ac),-log2(1-e6_epi_k27ac),-log2(1-e6_ve_k27ac), na.rm=TRUE)"
    ),
    iterator = 50, intervals = gintervals.all() %>% gintervals.diff(meth_dips) %>% gintervals.diff(k4me3_peaks) %>% gintervals.diff(k27me3_peaks) %>% gintervals.diff(get_promoters()), colnames = c("k4me1", "k27ac")
) %>%
    select(-intervalID) %>%
    as_tibble() %cache_df% here("output/chip_bg.tsv") %>%
    as_tibble()
```


### Figure S6F


We now compare the distribution of k4me1 and k27ac (enhancer markers) at methylation dips regions vs the background:

```{r}
options(repr.plot.width = 14, repr.plot.height = 4)
p_k4me1 <- meth_dips_chip %>%
    ggplot(aes(x = k4me1, y = 1 - ..y..)) +
    stat_ecdf() +
    stat_ecdf(data = chip_bg, color = "darkgray") +
    ylab("% loci") +
    scale_y_continuous(label = scales::percent)

p_k27ac <- meth_dips_chip %>%
    ggplot(aes(x = k27ac, y = 1 - ..y..)) +
    stat_ecdf() +
    stat_ecdf(data = chip_bg, color = "darkgray") +
    ylab("% loci") +
    scale_y_continuous(label = scales::percent)


p_k4me1 + p_k27ac
```

### Enhancer methylation distribution (Figure 5H, Figure S6G)

```{r}
d_bg_comp <- extract_meth_at_context(bg_intervs, iterator = "intervs.global.seq_CG", min_cov = 10) %cache_df% here("output/dips_bg_meth_comp.tsv") %>% as_tibble()
```

```{r}
d_bg_comp %>%
    filter(ctrl.cov >= 10, tet_tko.cov >= 10) %>%
    nrow()
```

```{r}
d_bg_comp %>%
    filter(e7.5.cov >= 10, tet_tko.cov >= 10) %>%
    nrow()
```

```{r}
options(repr.plot.width = 7, repr.plot.height = 7)
min_cov <- 10
plot_smooth_cpg_scatter(d_bg_comp, "ctrl", "tet_tko", min_cov, "Control", "Tet-TKO", colors = c("white", "lightgray", "lightblue", "blue", "red", "yellow"))
plot_smooth_cpg_scatter(d_bg_comp, "e7.5", "tet_tko", min_cov, "WT", "Tet-TKO", colors = c("white", "lightgray", "lightblue", "blue", "red", "yellow"))
```

```{r}
options(repr.plot.width = 5, repr.plot.height = 6)
plot_smooth_legend(c("white", "lightgray", "lightblue", "blue", "red", "yellow"))
```

### Distribution of DNA methylation around the center of putative enhancers (Figure 5I)

```{r}
df_dips_cpgs <- meth_df_comp_non_smoothed %>%
    select(chrom:end, starts_with("ctrl"), starts_with("tet_tko")) %>%
    gintervals.neighbors1(d_bg_comp %>% select(chrom, start, end, ctrl_dip = ctrl, wt_tet_tko = tet_tko) %>% mutate(strand = 1) %>% select(chrom, start, end, strand, everything()))
```

```{r}
options(repr.plot.width = 14, repr.plot.height = 7)
df <- df_dips_cpgs %>%
    filter(abs(dist) <= 5e4) %>%
    mutate(dist = cut(dist, breaks = seq(-500, 500, length.out = 6))) %>%
    filter(!is.na(dist)) %>%
    mutate(ctrl_dip = cut(ctrl_dip, c(0, 0.3, 0.7, 1), include.lowest = TRUE)) %>%
    group_by(chrom1, start1, end1, ctrl_dip, dist) %>%
    summarise_at(vars(ends_with(".cov"), ends_with(".meth")), sum, na.rm = TRUE) %>%
    ungroup() %>%
    mutate(Ctrl = ctrl.meth / ctrl.cov, TKO = tet_tko.meth / tet_tko.cov) %>%
    filter(ctrl.cov >= 10, tet_tko.cov >= 10)
df %>% count(ctrl_dip)
p_dips_boxp <- df %>%
    pivot_longer(Ctrl:TKO, names_to = "type", values_to = "meth") %>%
    ggplot(aes(x = dist, y = meth, fill = type)) +
    geom_boxplot(lwd = 0.01, outlier.size = 0.001, fatten = 0.1) +
    facet_grid(. ~ ctrl_dip) +
    scale_fill_manual(name = "", values = c("Ctrl" = "#00468BFF", "TKO" = "#ED0000FF")) +
    ylim(0, 1) +
    xlab("") +
    ylab("Methylation") +
    vertical_labs()
p_dips_boxp
```

```{r}
df %>%
    filter(dist == "(-100,100]") %>%
    group_by(ctrl_dip) %>%
    summarise_at(vars(Ctrl:TKO), mean, na.rm = TRUE) %>%
    mutate(diff = TKO - Ctrl) %>%
    mutate_at(vars(Ctrl:TKO), scales::percent)


df %>%
    filter(dist == "(-100,100]") %>%
    group_by(ctrl_dip) %>%
    summarise(p = ks.test(Ctrl, TKO)$p.value)
```

## Motifs (Figure S6H)


Due to the large size of the motif tracks, we provide here the script together with a pre-computed file:

```{r}
dips_bg_expanded <- d_bg_comp %>%
    distinct(chrom, start, end) %>%
    mutate(start = start - 150, end = end + 150)
# motifs <- {
#     library(gpwm)
#     gpwm.extract_all("jolma_10bp", dips_bg_expanded, parallel=FALSE)
#     }

motifs <- fread(here("data/dips_bg_expanded_motifs_jolma.tsv")) %>% as_tibble()
```

```{r}
motifs_df <- motifs %>%
    gather("motif", "pwm", -(chrom:end)) %>%
    filter(grepl("di_DBD", motif)) %>%
    mutate(motif = gsub("jolma_10bp\\.", "", motif)) %>%
    mutate(motif = gsub("_di_DBD", "", motif)) %>%
    as_tibble()
```

```{r}
motifs_df <- gextract_meth(comb_tracks, names = comb_names, min_cov = min_cov, intervals = motifs_df, iterator = motifs_df, join_intervals = TRUE) %>%
    select(-(chrom1:end1)) %>%
    rename(control = ctrl, control.cov = ctrl.cov)
```

Add global quantiles:

```{r}
# global_quantiles <- gpwm.get_global_quantiles("jolma_10bp", size=10)
global_quantiles <- fread(here("data/motifs_global_quantiles.tsv")) %>% as_tibble()
```

```{r}
motifs_df_thresh <- global_quantiles %>%
    filter(grepl("di_DBD", track)) %>%
    mutate(motif = gsub("_di_DBD", "", track)) %>%
    filter(quant == 0.995) %>%
    select(motif, thresh = value) %>%
    right_join(motifs_df) %>%
    mutate(hit = pwm >= thresh)
```

```{r}
motifs_summary <- motifs_df_thresh %>%
    group_by(motif) %>%
    mutate(diff = tet_tko - control) %>%
    mutate(p.value = suppressWarnings(ks.test(diff[hit], diff[!hit])$p.value)) %>%
    group_by(motif, hit) %>%
    summarise(p.value = mean(p.value), n = n(), control = mean(control, na.rm = TRUE), tet_tko = mean(tet_tko, na.rm = TRUE)) %>%
    ungroup() %>%
    mutate(hit = ifelse(hit, "on", "off")) %>%
    pivot_wider(names_from = "hit", values_from = c("n", "control", "tet_tko")) %>%
    select(motif, n_off, control_off, tet_tko_off, n_on, control_on, tet_tko_on, `pval (of on diff vs off diff)` = p.value)
```


### Table S3

```{r}
writexl::write_xlsx(motifs_summary, here("output/motifs_summary.xlsx"))
```

```{r}
motifs_to_plot <- c("POU2F3", "FOXJ3_4", "SOX11", "SOX1", "TCFAP2A", "IRX3", "SOX17", "HNF4A", "THRB", "FOXC1", "FOXC2", "MSX1")
motifs_to_plot <- map_chr(motifs_to_plot, ~ motifs_df_thresh %>%
    distinct(motif) %>%
    filter(grepl(.x, motif)) %>%
    slice(1) %>%
    pull(motif))
motifs_to_plot
```

```{r}
p_motifs_boxp <- motifs_df_thresh %>%
    filter(motif %in% motifs_to_plot) %>%
    select(-ends_with("cov")) %>%
    rename(Ctrl = control, TKO = tet_tko) %>%
    pivot_longer(names_to = "type", cols = c("Ctrl", "TKO"), values_to = "meth") %>%
    mutate(hit = ifelse(hit, "motif hotspot", "other enhancers")) %>%
    ggplot(aes(x = type, fill = hit, y = meth)) +
    geom_boxplot() +
    ylab("Methylation") +
    xlab("") +
    scale_fill_manual(name = "", values = rev(c("darkgray", "darkgreen"))) +
    facet_grid(. ~ motif)
p_motifs_boxp
```

<!--chapter:end:Methylation.Rmd-->

---
jupyter:
  jupytext:
    formats: ipynb,Rmd
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.4
  kernelspec:
    display_name: R
    language: R
    name: ir353
---

```{r, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
setwd(here::here())
```

```{r}
library("metacell")
library("tgstat")
library("tgconfig")
library("Matrix")
library("dplyr")
```

```{r}
scdb_init("scrna_db/", force_reinit = T)
```

```{r}
mat <- scdb_mat("tko_chim")
```

#### Remove cells belonging to E8.5 chimera embryos. Only use chimera embryos from E7.5

```{r}
e85_embryos <- paste0("TKO29_", c(13:19))
ig_cells <- colnames(mat@mat)[mat@cell_metadata[colnames(mat@mat), "embryo"] %in% e85_embryos]
ig_cells <- c(mat@ignore_cells, ig_cells)

mat <- scm_ignore_cells(scmat = mat, ig_cells = ig_cells)
```

#### Remove extraembryonic ectoderm and parietal endoderm cells

```{r}
source("scripts/pipeline/remove_exe_ectoderm_and_parietal_endo_cls.R")
mat <- remove_exe_ectoderm_and_parietal_endoderm(mat_query = mat)
scdb_add_mat(id = "tko_chim", mat = mat)
```

#### Merge cells with the wildtype atlas

```{r}
source("scripts/pipeline/merge_umi_mat_with_wt10_umi_mat.R")
mat_new <- merge_umi_mat_with_wt10(scmat = mat)
scdb_add_mat(id = "tko_chim_wt10", mat = mat_new)
```

#### Generate single-cell balanced kNN-graph
This step needs the file  
_data/tet_tko.bad_genes.txt_

```{r}
source("scripts/pipeline/gen_cgraph.R")
gen_cgraph(mat_nm = "tko_chim_wt10", force_recompute_cgraph = F)
```

#### Transfer cell type annotation to chimera cells

```{r}
source("scripts/pipeline/transfer_cell_type_annotation.R")
transfer_color_chimera_tetraploid(mat_nm = "tko_chim_wt10")
```

#### Time chimera embryos based on their host and KO cells

```{r}
source("scripts/pipeline/tko_chim_timing.R")
source("scripts/pipeline/transfer_time_annotation.R")
tko_chim_timing()
# tko_chimera_plot_cor_time_dist_per_emb(mat_nm = "tko_chim_wt10")
```

#### Export infered time and cell type into table

```{r}
source("scripts/pipeline/create_metadata.R")
create_tko_chimera_metadata()
```

#### Generate a separate UMI matrix and metacell object for the TKO cells

```{r}
source("scripts/pipeline/generat_ko_mat.R")
generate_ko_mat_chimera()
```

```{r}
source("scripts/pipeline/generic_mc.R")
tgconfig::override_params(config_file = "config/tet_chim.yaml", package = "metacell")

bad_genes <- read.table("data/tet_tko.bad_genes.txt", sep = "\t", stringsAsFactors = F)
bad_genes <- bad_genes[, 1]

mat_nm <- "tko_chim_ko"

generate_mc(name = mat_nm, add_bad_genes = bad_genes)
```

```{r}
wt_atlas <- mcell_gen_atlas(
    mat_id = "sing_emb_wt10",
    mc_id = "sing_emb_wt10_recolored",
    gset_id = "sing_emb_wt10",
    mc2d_id = "sing_emb_wt10_recolored"
)

cmp <- mcell_proj_on_atlas(
    mat_id = "tko_chim_ko",
    mc_id = "tko_chim_ko",
    atlas = wt_atlas,
    fig_cmp_dir = paste0("figs/atlas_projection_", "tko_chim_ko"), recolor_mc_id = "tko_chim_ko", ten2mars = F
)
```

```{r}
tgconfig::set_param(param = "mcell_mc2d_T_edge", value = 0.03, package = "metacell")
mcell_mc2d_force_knn(mc2d_id = "tko_chim_ko", mc_id = "tko_chim_ko", graph_id = "tko_chim_ko", graph_parametric = T, feats_gset = "tko_chim_ko")
mcell_mc2d_plot("tko_chim_ko")
```

```{r}

```

<!--chapter:end:TKO-chimera-embryos-processing.Rmd-->

---
jupyter:
  jupytext:
    formats: ipynb,Rmd
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.4
  kernelspec:
    display_name: R
    language: R
    name: ir353
---

```{r, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
setwd(here::here())
```

```{r}
library("metacell")
library("tgstat")
library("tgconfig")
library("Matrix")
library("dplyr")
```

```{r}
scdb_init("scrna_db/", force_reinit = T)
```

```{r}
mat <- scdb_mat("tko_tetra")
```

#### Remove extraembryonic ectoderm and parietal endoderm cells

```{r}
source("scripts/pipeline/remove_exe_ectoderm_and_parietal_endo_cls.R")
mat <- remove_exe_ectoderm_and_parietal_endoderm(mat_query = mat)
```

#### Merge cells with the wildtype atlas

```{r}
source("scripts/pipeline/merge_umi_mat_with_wt10_umi_mat.R")
mat_new <- merge_umi_mat_with_wt10(scmat = mat)
scdb_add_mat(id = "tko_tetra_wt10", mat = mat_new)
```

#### Generate single-cell balanced kNN-graph
This step needs the file  
_data/tet_tko.bad_genes.txt_

```{r}
source("scripts/pipeline/gen_cgraph.R")
gen_cgraph(mat_nm = "tko_tetra_wt10", force_recompute_cgraph = F)
```

#### Transfer cell type annotation to TKO cells

```{r}
source("scripts/pipeline/transfer_cell_type_annotation.R")
transfer_color_chimera_tetraploid(mat_nm = "tko_tetra_wt10")
```

#### Time tetraploid-complemented embryos based on their injected TKO cells

```{r}
source("scripts/pipeline/transfer_time_annotation.R")
source("scripts/pipeline/tetraploid_timing.R")
tetra_timing("tko_tetra_wt10")
```

#### Export infered time and cell type into table

```{r}
source("scripts/pipeline/create_metadata.R")
create_tko_tetraploid_metadata()
```

#### Generate a separate UMI matrix and metacell object for the TKO cells

```{r}
source("scripts/pipeline/generat_ko_mat.R")
generate_ko_mat_tetraploid()
```

```{r}
source("scripts/pipeline/generic_mc.R")
tgconfig::override_params(config_file = "config/tet_chim.yaml", package = "metacell")

bad_genes <- read.table("data/tet_tko.bad_genes.txt", sep = "\t", stringsAsFactors = F)
bad_genes <- bad_genes[, 1]

mat_nm <- "tko_tetra_ko"

generate_mc(mat_nm, add_bad_genes = bad_genes)
```

```{r}
wt_atlas <- mcell_gen_atlas(
    mat_id = "sing_emb_wt10",
    mc_id = "sing_emb_wt10_recolored",
    gset_id = "sing_emb_wt10",
    mc2d_id = "sing_emb_wt10_recolored"
)

cmp <- mcell_proj_on_atlas(
    mat_id = "tko_tetra_ko",
    mc_id = "tko_tetra_ko",
    atlas = wt_atlas,
    fig_cmp_dir = paste0("figs/atlas_projection_", "tko_tetra_ko"), recolor_mc_id = "tko_tetra_ko", ten2mars = F
)
```

```{r}
tgconfig::set_param(param = "mcell_mc2d_T_edge", value = 0.03, package = "metacell")
mcell_mc2d_force_knn(mc2d_id = "tko_tetra_ko", mc_id = "tko_tetra_ko", graph_id = "tko_tetra_ko", graph_parametric = T, feats_gset = "tko_tetra_ko")
mcell_mc2d_plot("tko_tetra_ko")
```

```{r}

```

```{r}

```

<!--chapter:end:TKO-tetraploid-complemented-embryos-processing.Rmd-->

