<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>2 Import scRNA-seq plates | The intrinsic and extrinsic effects of Tet proteins during gastrulation</title>
  <meta name="description" content="Analysis and code for the paper: The intrinsic and extrinsic effects of Tet proteins during gastrulation" />
  <meta name="generator" content="bookdown 0.26 and GitBook 2.6.7" />

  <meta property="og:title" content="2 Import scRNA-seq plates | The intrinsic and extrinsic effects of Tet proteins during gastrulation" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Analysis and code for the paper: The intrinsic and extrinsic effects of Tet proteins during gastrulation" />
  <meta name="github-repo" content="tanaylab/tet-gastrulation" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="2 Import scRNA-seq plates | The intrinsic and extrinsic effects of Tet proteins during gastrulation" />
  
  <meta name="twitter:description" content="Analysis and code for the paper: The intrinsic and extrinsic effects of Tet proteins during gastrulation" />
  

<meta name="author" content="Markus Mittnenzweig &amp; Aviezer Lifshitz" />


<meta name="date" content="2022-06-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="facs-gating-of-cell-population.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> The intrinsic and extrinsic effects of Tet proteins during gastrulation</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#run-the-notebooks"><i class="fa fa-check"></i><b>1.1</b> Run the notebooks</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#necessary-r-packages"><i class="fa fa-check"></i><b>1.2</b> Necessary R packages</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#notebook-order"><i class="fa fa-check"></i><b>1.3</b> Notebook order</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="import-scrna-seq-plates.html"><a href="import-scrna-seq-plates.html"><i class="fa fa-check"></i><b>2</b> Import scRNA-seq plates</a><ul>
<li class="chapter" data-level="2.0.1" data-path="import-scrna-seq-plates.html"><a href="import-scrna-seq-plates.html#import-tet-tko-chimera-plates-into-metacell"><i class="fa fa-check"></i><b>2.0.1</b> Import Tet TKO chimera plates into metacell</a></li>
<li class="chapter" data-level="2.0.2" data-path="import-scrna-seq-plates.html"><a href="import-scrna-seq-plates.html#import-tet-tko-tetraploid-complemented-embryos"><i class="fa fa-check"></i><b>2.0.2</b> Import Tet TKO tetraploid-complemented embryos</a></li>
<li class="chapter" data-level="2.0.3" data-path="import-scrna-seq-plates.html"><a href="import-scrna-seq-plates.html#import-control-tetraploid-complemented-embryos"><i class="fa fa-check"></i><b>2.0.3</b> Import Control tetraploid-complemented embryos</a></li>
<li class="chapter" data-level="2.0.4" data-path="import-scrna-seq-plates.html"><a href="import-scrna-seq-plates.html#import-tet-12-dko-chimera-embryos"><i class="fa fa-check"></i><b>2.0.4</b> Import Tet 1/2 DKO chimera embryos</a></li>
<li class="chapter" data-level="2.0.5" data-path="import-scrna-seq-plates.html"><a href="import-scrna-seq-plates.html#import-tet-13-dko-chimera-embryos"><i class="fa fa-check"></i><b>2.0.5</b> Import Tet 1/3 DKO chimera embryos</a></li>
<li class="chapter" data-level="2.0.6" data-path="import-scrna-seq-plates.html"><a href="import-scrna-seq-plates.html#import-tet-23-dko-chimera-embryos"><i class="fa fa-check"></i><b>2.0.6</b> Import Tet 2/3 DKO chimera embryos</a></li>
<li class="chapter" data-level="2.0.7" data-path="import-scrna-seq-plates.html"><a href="import-scrna-seq-plates.html#import-aav-cre-delivered-whole-embryo-tkos"><i class="fa fa-check"></i><b>2.0.7</b> Import AAV-Cre delivered whole-embryo TKOs</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="facs-gating-of-cell-population.html"><a href="facs-gating-of-cell-population.html"><i class="fa fa-check"></i><b>3</b> FACS gating of cell population</a></li>
<li class="chapter" data-level="4" data-path="tko-chimera-processing.html"><a href="tko-chimera-processing.html"><i class="fa fa-check"></i><b>4</b> TKO chimera processing</a></li>
<li class="chapter" data-level="5" data-path="processing-of-tetraploid-complemented-tko-embryos.html"><a href="processing-of-tetraploid-complemented-tko-embryos.html"><i class="fa fa-check"></i><b>5</b> Processing of tetraploid-complemented TKO embryos</a></li>
<li class="chapter" data-level="6" data-path="processing-of-tetraploid-complemented-control-embryos.html"><a href="processing-of-tetraploid-complemented-control-embryos.html"><i class="fa fa-check"></i><b>6</b> Processing of tetraploid-complemented control embryos</a></li>
<li class="chapter" data-level="7" data-path="dko-chimera-processing.html"><a href="dko-chimera-processing.html"><i class="fa fa-check"></i><b>7</b> DKO chimera processing</a></li>
<li class="chapter" data-level="8" data-path="processing-of-whole-embryo-tet-tko-embryos.html"><a href="processing-of-whole-embryo-tet-tko-embryos.html"><i class="fa fa-check"></i><b>8</b> Processing of whole-embryo Tet TKO embryos</a></li>
<li class="chapter" data-level="9" data-path="methylation-analysis.html"><a href="methylation-analysis.html"><i class="fa fa-check"></i><b>9</b> Methylation analysis</a><ul>
<li class="chapter" data-level="9.1" data-path="methylation-analysis.html"><a href="methylation-analysis.html#initialization"><i class="fa fa-check"></i><b>9.1</b> Initialization</a></li>
<li class="chapter" data-level="9.2" data-path="methylation-analysis.html"><a href="methylation-analysis.html#define-genomic-contexts"><i class="fa fa-check"></i><b>9.2</b> Define genomic contexts</a></li>
<li class="chapter" data-level="9.3" data-path="methylation-analysis.html"><a href="methylation-analysis.html#extract-methylation-data"><i class="fa fa-check"></i><b>9.3</b> Extract methylation data</a><ul>
<li class="chapter" data-level="9.3.1" data-path="methylation-analysis.html"><a href="methylation-analysis.html#all-cpgs"><i class="fa fa-check"></i><b>9.3.1</b> All CpGs</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="methylation-analysis.html"><a href="methylation-analysis.html#distribution-of-tet-tko-vs-wt-methylation-figure-5b"><i class="fa fa-check"></i><b>9.4</b> Distribution of TET-TKO vs WT methylation (Figure 5B)</a></li>
<li class="chapter" data-level="9.5" data-path="methylation-analysis.html"><a href="methylation-analysis.html#methylation-in-tads"><i class="fa fa-check"></i><b>9.5</b> Methylation in TADs</a><ul>
<li class="chapter" data-level="9.5.1" data-path="methylation-analysis.html"><a href="methylation-analysis.html#compare-tad-methylation-in-ctrl-and-tet-tko-figure-5c"><i class="fa fa-check"></i><b>9.5.1</b> Compare TAD methylation in Ctrl and TET-tko (Figure 5C)</a></li>
<li class="chapter" data-level="9.5.2" data-path="methylation-analysis.html"><a href="methylation-analysis.html#figure-s6a"><i class="fa fa-check"></i><b>9.5.2</b> Figure S6A</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="methylation-analysis.html"><a href="methylation-analysis.html#tad-methylation-at-different-tor-regimes-figure-5d"><i class="fa fa-check"></i><b>9.6</b> TAD methylation at different TOR regimes (Figure 5D)</a></li>
<li class="chapter" data-level="9.7" data-path="methylation-analysis.html"><a href="methylation-analysis.html#low-methylation-hotspots"><i class="fa fa-check"></i><b>9.7</b> Low methylation hotspots</a><ul>
<li class="chapter" data-level="9.7.1" data-path="methylation-analysis.html"><a href="methylation-analysis.html#figure-s6b"><i class="fa fa-check"></i><b>9.7.1</b> Figure S6B</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="methylation-analysis.html"><a href="methylation-analysis.html#k4me3k27me3-hospots-figure-5e"><i class="fa fa-check"></i><b>9.8</b> K4me3/K27me3 hospots (Figure 5E)</a><ul>
<li class="chapter" data-level="9.8.1" data-path="methylation-analysis.html"><a href="methylation-analysis.html#k4me3-hotspots"><i class="fa fa-check"></i><b>9.8.1</b> K4me3 hotspots</a></li>
<li class="chapter" data-level="9.8.2" data-path="methylation-analysis.html"><a href="methylation-analysis.html#k27me3-hotspots"><i class="fa fa-check"></i><b>9.8.2</b> K27me3 hotspots</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="methylation-analysis.html"><a href="methylation-analysis.html#loci-linked-with-ctcf-occupancy-figure-5f"><i class="fa fa-check"></i><b>9.9</b> Loci linked with CTCF occupancy (Figure 5F)</a></li>
<li class="chapter" data-level="9.10" data-path="methylation-analysis.html"><a href="methylation-analysis.html#methylation-at-exons-and-non-h3k4me3-promoters-figure-5g"><i class="fa fa-check"></i><b>9.10</b> Methylation at Exons and Non-H3K4me3 Promoters (Figure 5G)</a><ul>
<li class="chapter" data-level="9.10.1" data-path="methylation-analysis.html"><a href="methylation-analysis.html#exons"><i class="fa fa-check"></i><b>9.10.1</b> Exons</a></li>
<li class="chapter" data-level="9.10.2" data-path="methylation-analysis.html"><a href="methylation-analysis.html#compare-with-the-wt-figure-s6c"><i class="fa fa-check"></i><b>9.10.2</b> Compare with the WT (Figure S6C):</a></li>
<li class="chapter" data-level="9.10.3" data-path="methylation-analysis.html"><a href="methylation-analysis.html#promoters"><i class="fa fa-check"></i><b>9.10.3</b> Promoters</a></li>
<li class="chapter" data-level="9.10.4" data-path="methylation-analysis.html"><a href="methylation-analysis.html#compare-with-the-wt-figure-s6c-1"><i class="fa fa-check"></i><b>9.10.4</b> Compare with the WT (Figure S6C):</a></li>
</ul></li>
<li class="chapter" data-level="9.11" data-path="methylation-analysis.html"><a href="methylation-analysis.html#promoter-methylation-vs-expression-figure-s6ds6e"><i class="fa fa-check"></i><b>9.11</b> Promoter methylation vs expression (Figure S6D/S6E)</a></li>
<li class="chapter" data-level="9.12" data-path="methylation-analysis.html"><a href="methylation-analysis.html#putative-enhancers"><i class="fa fa-check"></i><b>9.12</b> Putative enhancers</a><ul>
<li class="chapter" data-level="9.12.1" data-path="methylation-analysis.html"><a href="methylation-analysis.html#figure-s6f"><i class="fa fa-check"></i><b>9.12.1</b> Figure S6F</a></li>
<li class="chapter" data-level="9.12.2" data-path="methylation-analysis.html"><a href="methylation-analysis.html#enhancer-methylation-distribution-figure-5h-figure-s6g"><i class="fa fa-check"></i><b>9.12.2</b> Enhancer methylation distribution (Figure 5H, Figure S6G)</a></li>
<li class="chapter" data-level="9.12.3" data-path="methylation-analysis.html"><a href="methylation-analysis.html#distribution-of-dna-methylation-around-the-center-of-putative-enhancers-figure-5i"><i class="fa fa-check"></i><b>9.12.3</b> Distribution of DNA methylation around the center of putative enhancers (Figure 5I)</a></li>
</ul></li>
<li class="chapter" data-level="9.13" data-path="methylation-analysis.html"><a href="methylation-analysis.html#motifs-figure-s6h"><i class="fa fa-check"></i><b>9.13</b> Motifs (Figure S6H)</a><ul>
<li class="chapter" data-level="9.13.1" data-path="methylation-analysis.html"><a href="methylation-analysis.html#table-s3"><i class="fa fa-check"></i><b>9.13.1</b> Table S3</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="generate-figures-and-supplementary-figures.html"><a href="generate-figures-and-supplementary-figures.html"><i class="fa fa-check"></i><b>10</b> Generate figures and supplementary figures</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The intrinsic and extrinsic effects of Tet proteins during gastrulation</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="import-scrna-seq-plates" class="section level1 hasAnchor">
<h1><span class="header-section-number">2</span> Import scRNA-seq plates<a href="import-scrna-seq-plates.html#import-scrna-seq-plates" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="import-scrna-seq-plates.html#cb3-1"></a><span class="kw">setwd</span>(here<span class="op">::</span><span class="kw">here</span>())</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="import-scrna-seq-plates.html#cb4-1"></a><span class="kw">library</span>(<span class="st">&quot;metacell&quot;</span>)</span>
<span id="cb4-2"><a href="import-scrna-seq-plates.html#cb4-2"></a><span class="kw">library</span>(<span class="st">&quot;Matrix&quot;</span>)</span>
<span id="cb4-3"><a href="import-scrna-seq-plates.html#cb4-3"></a><span class="kw">library</span>(<span class="st">&quot;dplyr&quot;</span>)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="import-scrna-seq-plates.html#cb8-1"></a><span class="kw">scdb_init</span>(<span class="st">&quot;scrna_db/&quot;</span>, <span class="dt">force_reinit =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## initializing scdb to scrna_db/</code></pre>
<div id="import-tet-tko-chimera-plates-into-metacell" class="section level3 hasAnchor">
<h3><span class="header-section-number">2.0.1</span> Import Tet TKO chimera plates into metacell<a href="import-scrna-seq-plates.html#import-tet-tko-chimera-plates-into-metacell" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This notebook requires that the UMI tables for each plate are available in the folder <strong>data/umi.tables</strong> . If they are not available, please run the cell below to download the necessary data</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="import-scrna-seq-plates.html#cb10-1"></a><span class="co"># If you need to download the umi tables, please run the two lines below</span></span>
<span id="cb10-2"><a href="import-scrna-seq-plates.html#cb10-2"></a><span class="cf">if</span> (<span class="dv">0</span>) {</span>
<span id="cb10-3"><a href="import-scrna-seq-plates.html#cb10-3"></a>    <span class="kw">source</span>(<span class="st">&quot;../scripts/download_data.R&quot;</span>)</span>
<span id="cb10-4"><a href="import-scrna-seq-plates.html#cb10-4"></a>    <span class="kw">download_umi_tables</span>()</span>
<span id="cb10-5"><a href="import-scrna-seq-plates.html#cb10-5"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="import-scrna-seq-plates.html#cb11-1"></a><span class="kw">source</span>(<span class="st">&quot;scripts/pipeline/import_plates.R&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="import-scrna-seq-plates.html#cb12-1"></a>mat_nm &lt;-<span class="st"> &quot;tko_chim&quot;</span></span>
<span id="cb12-2"><a href="import-scrna-seq-plates.html#cb12-2"></a></span>
<span id="cb12-3"><a href="import-scrna-seq-plates.html#cb12-3"></a>metadata &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&quot;data/plate_metadata.csv&quot;</span>, <span class="dt">stringsAsFactors =</span> F)</span>
<span id="cb12-4"><a href="import-scrna-seq-plates.html#cb12-4"></a>genotype &lt;-<span class="st"> &quot;TKO&quot;</span></span>
<span id="cb12-5"><a href="import-scrna-seq-plates.html#cb12-5"></a>treatment &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Chimera assay&quot;</span>)</span>
<span id="cb12-6"><a href="import-scrna-seq-plates.html#cb12-6"></a>metadata &lt;-<span class="st"> </span>metadata[metadata<span class="op">$</span>treatment <span class="op">%in%</span><span class="st"> </span>treatment, ]</span>
<span id="cb12-7"><a href="import-scrna-seq-plates.html#cb12-7"></a>metadata &lt;-<span class="st"> </span>metadata[<span class="kw">grep</span>(<span class="dt">pattern =</span> genotype, <span class="dt">x =</span> metadata<span class="op">$</span>genotype), ]</span>
<span id="cb12-8"><a href="import-scrna-seq-plates.html#cb12-8"></a></span>
<span id="cb12-9"><a href="import-scrna-seq-plates.html#cb12-9"></a><span class="co"># remove additionally the plate 190730_P03. No embryo and facs data available. Only relevant for TKO Chimera</span></span>
<span id="cb12-10"><a href="import-scrna-seq-plates.html#cb12-10"></a>metadata &lt;-<span class="st"> </span>metadata[metadata<span class="op">$</span>plate <span class="op">!=</span><span class="st"> &quot;190730_P03&quot;</span>, ]</span>
<span id="cb12-11"><a href="import-scrna-seq-plates.html#cb12-11"></a></span>
<span id="cb12-12"><a href="import-scrna-seq-plates.html#cb12-12"></a><span class="kw">import_plates</span>(<span class="dt">mat_nm =</span> mat_nm, <span class="dt">metadata =</span> metadata)</span></code></pre></div>
<pre><code>## will read 190303_P02</code></pre>
<pre><code>## will read 190318_P01</code></pre>
<pre><code>## will read 190318_P02</code></pre>
<pre><code>## will read 190318_P03</code></pre>
<pre><code>## will read 190318_P04</code></pre>
<pre><code>## will read 190318_P05</code></pre>
<pre><code>## will read 190318_P06</code></pre>
<pre><code>## will read 190318_P07</code></pre>
<pre><code>## will read 190318_P09</code></pre>
<pre><code>## will read 190318_P11</code></pre>
<pre><code>## will read 190318_P12</code></pre>
<pre><code>## will read 190322_P01</code></pre>
<pre><code>## will read 190322_P02</code></pre>
<pre><code>## will read 190322_P04</code></pre>
<pre><code>## will read 190325_P01</code></pre>
<pre><code>## will read 190325_P02</code></pre>
<pre><code>## will read 190325_P03</code></pre>
<pre><code>## will read 190325_P04</code></pre>
<pre><code>## will read 190325_P05</code></pre>
<pre><code>## will read 190325_P07</code></pre>
<pre><code>## will read 190325_P08</code></pre>
<pre><code>## will read 190730_P01</code></pre>
<pre><code>## will read 190730_P02</code></pre>
<pre><code>## will read 190730_P04</code></pre>
<pre><code>## will read 190730_P05</code></pre>
<pre><code>## will read 190730_P06</code></pre>
<pre><code>## will read 190730_P07</code></pre>
<pre><code>## will read 190730_P08</code></pre>
<pre><code>## will read 190730_P09</code></pre>
<pre><code>## will read 190730_P10</code></pre>
<pre><code>## will read 190730_P11</code></pre>
<pre><code>## will read 190806_P04</code></pre>
<pre><code>## will read 190806_P05</code></pre>
<pre><code>## will read 190806_P06</code></pre>
<pre><code>## will read 20190806_P01</code></pre>
<pre><code>## will read 20190806_P02</code></pre>
<pre><code>## will read 20190806_P03</code></pre>
<pre><code>## will read 20190806_P07</code></pre>
<pre><code>## will read 20190806_P08</code></pre>
<pre><code>## will read 20190806_P09</code></pre>
<pre><code>## will read 1106TKO29_01</code></pre>
<pre><code>## will read 1106TKO29_02</code></pre>
<pre><code>## will read 1106TKO29_03</code></pre>
<pre><code>## will read 1106TKO29_04</code></pre>
<pre><code>## will read 1106TKO29_05</code></pre>
<pre><code>## will read 1106TKO29_06</code></pre>
<pre><code>## will read 1106TKO29_07</code></pre>
<pre><code>## will read 1106TKO29_08</code></pre>
<pre><code>## will read 1106TKO29_09</code></pre>
<pre><code>## will read 1106TKO29_10</code></pre>
<pre><code>## will read 1106TKO29_11</code></pre>
<pre><code>## will read 1106TKO29_12</code></pre>
<pre><code>## will read 1106TKO29_13</code></pre>
<pre><code>## will read 1106TKO29_14</code></pre>
<pre><code>## will read 1106TKO29_15</code></pre>
<pre><code>## will read 191108_P01</code></pre>
<pre><code>## will read 191108_P02</code></pre>
<pre><code>## will read 191108_P03</code></pre>
<pre><code>## will read 191108_P04</code></pre>
<pre><code>## will read 191108_P05</code></pre>
<pre><code>## will read 191108_P06</code></pre>
<pre><code>## will read 191108_P07</code></pre>
<pre><code>## will read 191108_P08</code></pre>
<pre><code>## will read 20200106_TKO26_P01_NOVA</code></pre>
<pre><code>## will read 20200106_TKO26_P02_NOVA</code></pre>
<pre><code>## will read 20200106_TKO26_P03_NOVA</code></pre>
<pre><code>## will read 20200106_TKO26_P04_NOVA</code></pre>
<pre><code>## will read 20200106_TKO26_P05_NOVA</code></pre>
<pre><code>## will read TKO26_13_P01</code></pre>
<pre><code>## will read TKO26_13_P02</code></pre>
<pre><code>## will read TKO26_13_P03</code></pre>
<pre><code>## will read TKO26_13_P04</code></pre>
<pre><code>## will read TKO26_13_P05</code></pre>
<pre><code>## will read TKO26_13_P06</code></pre>
<pre><code>## will read TKO26_13_P07</code></pre>
<pre><code>## will read TKO26_13_P08</code></pre>
<pre><code>## will read TKO26_13_P09</code></pre>
<pre><code>## will read TKO26_13_P10</code></pre>
<pre><code>## will read TKO26_13_P11</code></pre>
<pre><code>## will read TKO26_13_P12</code></pre>
<pre><code>## will read TKO26_13_P15</code></pre>
<pre><code>## will read TetTKO23_P01</code></pre>
<pre><code>## will read TetTKO23_P02</code></pre>
<pre><code>## will read TetTKO23_P03</code></pre>
<pre><code>## will read TetTKO23_P04</code></pre>
<pre><code>## will read TetTKO23_P05</code></pre>
<pre><code>## will read TetTKO23_P06</code></pre>
<pre><code>## will read TetTKO23_P07</code></pre>
<pre><code>## will read TetTKO23_P08</code></pre>
<pre><code>## will read TKO23_P01</code></pre>
<pre><code>## will read TKO23_P02</code></pre>
<pre><code>## will read TKO23_P04</code></pre>
<pre><code>## will read TKO23_P05</code></pre>
<pre><code>## will read TKO23_P06</code></pre>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="import-tet-tko-tetraploid-complemented-embryos" class="section level3 hasAnchor">
<h3><span class="header-section-number">2.0.2</span> Import Tet TKO tetraploid-complemented embryos<a href="import-scrna-seq-plates.html#import-tet-tko-tetraploid-complemented-embryos" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="import-scrna-seq-plates.html#cb108-1"></a>mat_nm &lt;-<span class="st"> &quot;tko_tetra&quot;</span></span>
<span id="cb108-2"><a href="import-scrna-seq-plates.html#cb108-2"></a></span>
<span id="cb108-3"><a href="import-scrna-seq-plates.html#cb108-3"></a>metadata &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&quot;data/plate_metadata.csv&quot;</span>, <span class="dt">stringsAsFactors =</span> F)</span>
<span id="cb108-4"><a href="import-scrna-seq-plates.html#cb108-4"></a>genotype &lt;-<span class="st"> &quot;TKO&quot;</span></span>
<span id="cb108-5"><a href="import-scrna-seq-plates.html#cb108-5"></a>treatment &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Tetraploid complementation assay&quot;</span>)</span>
<span id="cb108-6"><a href="import-scrna-seq-plates.html#cb108-6"></a>metadata &lt;-<span class="st"> </span>metadata[metadata<span class="op">$</span>treatment <span class="op">%in%</span><span class="st"> </span>treatment, ]</span>
<span id="cb108-7"><a href="import-scrna-seq-plates.html#cb108-7"></a>metadata &lt;-<span class="st"> </span>metadata[<span class="kw">grep</span>(<span class="dt">pattern =</span> genotype, <span class="dt">x =</span> metadata<span class="op">$</span>genotype), ]</span>
<span id="cb108-8"><a href="import-scrna-seq-plates.html#cb108-8"></a></span>
<span id="cb108-9"><a href="import-scrna-seq-plates.html#cb108-9"></a><span class="kw">import_plates</span>(<span class="dt">mat_nm =</span> mat_nm, <span class="dt">metadata =</span> metadata)</span></code></pre></div>
<pre><code>## will read TKO26_tetra_191111_P01_NOVAseq</code></pre>
<pre><code>## will read TKO26_tetra_191111_P02</code></pre>
<pre><code>## will read TKO26_tetra_191111_P03</code></pre>
<pre><code>## will read TKO26_tetra_191111_P04</code></pre>
<pre><code>## will read 191125_P01_NOVA</code></pre>
<pre><code>## will read 191125_P02</code></pre>
<pre><code>## will read 191125_P03</code></pre>
<pre><code>## will read 191125_P04</code></pre>
<pre><code>## will read 4N_P01</code></pre>
<pre><code>## will read 4N_P06</code></pre>
<pre><code>## will read 4N26_01</code></pre>
<pre><code>## will read 4N26_02</code></pre>
<pre><code>## will read 4N26_03</code></pre>
<pre><code>## will read 4N26_04</code></pre>
<pre><code>## will read 4N26_05</code></pre>
<pre><code>## will read 4N26_06</code></pre>
<pre><code>## will read 4N26_07</code></pre>
<pre><code>## will read 4N26_08</code></pre>
<pre><code>## will read 4N_P07</code></pre>
<pre><code>## will read 190819_P01</code></pre>
<pre><code>## will read 190819_P02</code></pre>
<pre><code>## will read 190819_P03</code></pre>
<pre><code>## will read 190819_P04</code></pre>
<pre><code>## will read 190916_P07</code></pre>
<pre><code>## will read 190916_P08</code></pre>
<pre><code>## will read 190916_P09</code></pre>
<pre><code>## will read 4N_P04</code></pre>
<pre><code>## will read 4N_P05</code></pre>
<pre><code>## will read 4N_P08</code></pre>
<pre><code>## will read 4N_P10</code></pre>
<pre><code>## will read 4N_P11</code></pre>
<pre><code>## will read 4N_P12</code></pre>
<pre><code>## will read 4N_P13</code></pre>
<pre><code>## will read 4N_P14</code></pre>
<pre><code>## will read 4N29_01</code></pre>
<pre><code>## will read 4N29_02</code></pre>
<pre><code>## will read 4N29_03</code></pre>
<pre><code>## will read 4N29_04</code></pre>
<pre><code>## will read 4N29_05</code></pre>
<pre><code>## will read 4N29_06</code></pre>
<pre><code>## will read 4N29_07</code></pre>
<pre><code>## will read 4N29_08</code></pre>
<pre><code>## will read 4N29_09</code></pre>
<pre><code>## will read 4N29_10</code></pre>
<pre><code>## will read 4N29_11</code></pre>
<pre><code>## will read 4N29_12</code></pre>
<pre><code>## will read 4N29_13</code></pre>
<pre><code>## will read 4N29_14</code></pre>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="import-scrna-seq-plates.html#cb158-1"></a><span class="co"># There are four mixed plates containing both tetraploid-complemented control and Tet TKO embryos</span></span>
<span id="cb158-2"><a href="import-scrna-seq-plates.html#cb158-2"></a><span class="co"># We additionally ignore the cells from control embryos</span></span>
<span id="cb158-3"><a href="import-scrna-seq-plates.html#cb158-3"></a>mat &lt;-<span class="st"> </span><span class="kw">scdb_mat</span>(<span class="st">&quot;tko_tetra&quot;</span>)</span>
<span id="cb158-4"><a href="import-scrna-seq-plates.html#cb158-4"></a></span>
<span id="cb158-5"><a href="import-scrna-seq-plates.html#cb158-5"></a>ctrl_cells &lt;-<span class="st"> </span>mat<span class="op">@</span>cell_metadata<span class="op">$</span>cell[<span class="kw">grep</span>(<span class="dt">pattern =</span> <span class="st">&quot;Ctrl&quot;</span>, <span class="dt">x =</span> mat<span class="op">@</span>cell_metadata<span class="op">$</span>embryo)]</span>
<span id="cb158-6"><a href="import-scrna-seq-plates.html#cb158-6"></a></span>
<span id="cb158-7"><a href="import-scrna-seq-plates.html#cb158-7"></a>mat &lt;-<span class="st"> </span><span class="kw">scm_ignore_cells</span>(<span class="dt">scmat =</span> mat, <span class="dt">ig_cells =</span> <span class="kw">union</span>(mat<span class="op">@</span>ignore_cells, ctrl_cells))</span>
<span id="cb158-8"><a href="import-scrna-seq-plates.html#cb158-8"></a></span>
<span id="cb158-9"><a href="import-scrna-seq-plates.html#cb158-9"></a><span class="kw">scdb_add_mat</span>(<span class="dt">id =</span> <span class="st">&quot;tko_tetra&quot;</span>, <span class="dt">mat =</span> mat)</span></code></pre></div>
</div>
<div id="import-control-tetraploid-complemented-embryos" class="section level3 hasAnchor">
<h3><span class="header-section-number">2.0.3</span> Import Control tetraploid-complemented embryos<a href="import-scrna-seq-plates.html#import-control-tetraploid-complemented-embryos" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="import-scrna-seq-plates.html#cb159-1"></a>mat_nm &lt;-<span class="st"> &quot;control_tetra&quot;</span></span>
<span id="cb159-2"><a href="import-scrna-seq-plates.html#cb159-2"></a></span>
<span id="cb159-3"><a href="import-scrna-seq-plates.html#cb159-3"></a>metadata &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&quot;data/plate_metadata.csv&quot;</span>, <span class="dt">stringsAsFactors =</span> F)</span>
<span id="cb159-4"><a href="import-scrna-seq-plates.html#cb159-4"></a>genotype &lt;-<span class="st"> &quot;Control&quot;</span></span>
<span id="cb159-5"><a href="import-scrna-seq-plates.html#cb159-5"></a>treatment &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Tetraploid complementation assay&quot;</span>)</span>
<span id="cb159-6"><a href="import-scrna-seq-plates.html#cb159-6"></a>metadata &lt;-<span class="st"> </span>metadata[metadata<span class="op">$</span>treatment <span class="op">%in%</span><span class="st"> </span>treatment, ]</span>
<span id="cb159-7"><a href="import-scrna-seq-plates.html#cb159-7"></a>metadata &lt;-<span class="st"> </span>metadata[<span class="kw">grep</span>(<span class="dt">pattern =</span> genotype, <span class="dt">x =</span> metadata<span class="op">$</span>genotype), ]</span>
<span id="cb159-8"><a href="import-scrna-seq-plates.html#cb159-8"></a></span>
<span id="cb159-9"><a href="import-scrna-seq-plates.html#cb159-9"></a><span class="kw">import_plates</span>(<span class="dt">mat_nm =</span> mat_nm, <span class="dt">metadata =</span> metadata)</span></code></pre></div>
<pre><code>## will read 191125_P04</code></pre>
<pre><code>## will read 190819_P01</code></pre>
<pre><code>## will read 190916_P07</code></pre>
<pre><code>## will read 190916_P09</code></pre>
<pre><code>## will read 190916_P01</code></pre>
<pre><code>## will read 190916_P02</code></pre>
<pre><code>## will read 190916_P03</code></pre>
<pre><code>## will read 190916_P04</code></pre>
<pre><code>## will read 190916_P05</code></pre>
<pre><code>## will read 190916_P06</code></pre>
<pre><code>## will read 0217_P01</code></pre>
<pre><code>## will read 0217_P02</code></pre>
<pre><code>## will read 0217_P03</code></pre>
<pre><code>## will read 0217_P04</code></pre>
<pre><code>## will read 0217_P05</code></pre>
<pre><code>## will read 191125_P05</code></pre>
<pre><code>## will read 191125_P06_NOVA</code></pre>
<pre><code>## will read 191125_P07</code></pre>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="import-scrna-seq-plates.html#cb179-1"></a><span class="co"># There are four mixed plates containing both tetraploid-complemented control and Tet TKO embryos</span></span>
<span id="cb179-2"><a href="import-scrna-seq-plates.html#cb179-2"></a><span class="co"># We additionally ignore the cells from TKO embryos</span></span>
<span id="cb179-3"><a href="import-scrna-seq-plates.html#cb179-3"></a>mat &lt;-<span class="st"> </span><span class="kw">scdb_mat</span>(<span class="st">&quot;control_tetra&quot;</span>)</span>
<span id="cb179-4"><a href="import-scrna-seq-plates.html#cb179-4"></a></span>
<span id="cb179-5"><a href="import-scrna-seq-plates.html#cb179-5"></a>tko_cells &lt;-<span class="st"> </span>mat<span class="op">@</span>cell_metadata<span class="op">$</span>cell[<span class="kw">grep</span>(<span class="dt">pattern =</span> <span class="st">&quot;TKO&quot;</span>, <span class="dt">x =</span> mat<span class="op">@</span>cell_metadata<span class="op">$</span>embryo)]</span>
<span id="cb179-6"><a href="import-scrna-seq-plates.html#cb179-6"></a><span class="kw">length</span>(tko_cells)</span></code></pre></div>
<pre><code>## [1] 809</code></pre>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="import-scrna-seq-plates.html#cb181-1"></a>mat &lt;-<span class="st"> </span><span class="kw">scm_ignore_cells</span>(<span class="dt">scmat =</span> mat, <span class="dt">ig_cells =</span> <span class="kw">union</span>(mat<span class="op">@</span>ignore_cells, tko_cells))</span>
<span id="cb181-2"><a href="import-scrna-seq-plates.html#cb181-2"></a></span>
<span id="cb181-3"><a href="import-scrna-seq-plates.html#cb181-3"></a><span class="kw">scdb_add_mat</span>(<span class="dt">id =</span> <span class="st">&quot;control_tetra&quot;</span>, <span class="dt">mat =</span> mat)</span></code></pre></div>
</div>
<div id="import-tet-12-dko-chimera-embryos" class="section level3 hasAnchor">
<h3><span class="header-section-number">2.0.4</span> Import Tet 1/2 DKO chimera embryos<a href="import-scrna-seq-plates.html#import-tet-12-dko-chimera-embryos" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="import-scrna-seq-plates.html#cb182-1"></a>mat_nm &lt;-<span class="st"> &quot;dko12_chim&quot;</span></span>
<span id="cb182-2"><a href="import-scrna-seq-plates.html#cb182-2"></a></span>
<span id="cb182-3"><a href="import-scrna-seq-plates.html#cb182-3"></a>metadata &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&quot;data/plate_metadata.csv&quot;</span>, <span class="dt">stringsAsFactors =</span> F)</span>
<span id="cb182-4"><a href="import-scrna-seq-plates.html#cb182-4"></a>genotype &lt;-<span class="st"> &quot;Tet1/2 DKO&quot;</span></span>
<span id="cb182-5"><a href="import-scrna-seq-plates.html#cb182-5"></a>treatment &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Chimera assay&quot;</span>)</span>
<span id="cb182-6"><a href="import-scrna-seq-plates.html#cb182-6"></a>metadata &lt;-<span class="st"> </span>metadata[metadata<span class="op">$</span>treatment <span class="op">%in%</span><span class="st"> </span>treatment, ]</span>
<span id="cb182-7"><a href="import-scrna-seq-plates.html#cb182-7"></a>metadata &lt;-<span class="st"> </span>metadata[<span class="kw">grep</span>(<span class="dt">pattern =</span> genotype, <span class="dt">x =</span> metadata<span class="op">$</span>genotype), ]</span>
<span id="cb182-8"><a href="import-scrna-seq-plates.html#cb182-8"></a></span>
<span id="cb182-9"><a href="import-scrna-seq-plates.html#cb182-9"></a><span class="kw">import_plates</span>(<span class="dt">mat_nm =</span> mat_nm, <span class="dt">metadata =</span> metadata)</span></code></pre></div>
<pre><code>## will read Tet12_0228_P01</code></pre>
<pre><code>## will read Tet12_0228_P02</code></pre>
<pre><code>## will read Tet12_0228_P03</code></pre>
<pre><code>## will read Tet12_0228_P04</code></pre>
<pre><code>## will read Tet12_0228_P05</code></pre>
<pre><code>## will read Tet12_0228_P06</code></pre>
<pre><code>## will read Tet12_0228_P07</code></pre>
<pre><code>## will read Tet12_0228_P08</code></pre>
<pre><code>## will read Tet12_0228_P09</code></pre>
<pre><code>## will read Tet12_0228_P10</code></pre>
<pre><code>## will read Tet12_0303_P01</code></pre>
<pre><code>## will read Tet12_0303_P02</code></pre>
<pre><code>## will read Tet12_0303_P03</code></pre>
<pre><code>## will read Tet12_0303_P04</code></pre>
<pre><code>## will read Tet12_1_0616</code></pre>
<pre><code>## will read Tet12_2_0616</code></pre>
<pre><code>## will read Tet12DKO_10</code></pre>
<pre><code>## will read Tet12DKO_11</code></pre>
<pre><code>## will read Tet12DKO_12</code></pre>
<pre><code>## will read Tet12DKO_13</code></pre>
<pre><code>## will read Tet12DKO_14</code></pre>
<pre><code>## will read Tet12DKO_15</code></pre>
<pre><code>## will read Tet12DKO_7</code></pre>
<pre><code>## will read Tet12DKO_8</code></pre>
<pre><code>## will read Tet12DKO_9</code></pre>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="import-tet-13-dko-chimera-embryos" class="section level3 hasAnchor">
<h3><span class="header-section-number">2.0.5</span> Import Tet 1/3 DKO chimera embryos<a href="import-scrna-seq-plates.html#import-tet-13-dko-chimera-embryos" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="import-scrna-seq-plates.html#cb209-1"></a>mat_nm &lt;-<span class="st"> &quot;dko13_chim&quot;</span></span>
<span id="cb209-2"><a href="import-scrna-seq-plates.html#cb209-2"></a></span>
<span id="cb209-3"><a href="import-scrna-seq-plates.html#cb209-3"></a>metadata &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&quot;data/plate_metadata.csv&quot;</span>, <span class="dt">stringsAsFactors =</span> F)</span>
<span id="cb209-4"><a href="import-scrna-seq-plates.html#cb209-4"></a>genotype &lt;-<span class="st"> &quot;Tet1/3 DKO&quot;</span></span>
<span id="cb209-5"><a href="import-scrna-seq-plates.html#cb209-5"></a>treatment &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Chimera assay&quot;</span>)</span>
<span id="cb209-6"><a href="import-scrna-seq-plates.html#cb209-6"></a>metadata &lt;-<span class="st"> </span>metadata[metadata<span class="op">$</span>treatment <span class="op">%in%</span><span class="st"> </span>treatment, ]</span>
<span id="cb209-7"><a href="import-scrna-seq-plates.html#cb209-7"></a>metadata &lt;-<span class="st"> </span>metadata[<span class="kw">grep</span>(<span class="dt">pattern =</span> genotype, <span class="dt">x =</span> metadata<span class="op">$</span>genotype), ]</span>
<span id="cb209-8"><a href="import-scrna-seq-plates.html#cb209-8"></a></span>
<span id="cb209-9"><a href="import-scrna-seq-plates.html#cb209-9"></a><span class="kw">import_plates</span>(<span class="dt">mat_nm =</span> mat_nm, <span class="dt">metadata =</span> metadata)</span></code></pre></div>
<pre><code>## will read mTet13_P01</code></pre>
<pre><code>## will read mTet13_P03</code></pre>
<pre><code>## will read mTet13_P07</code></pre>
<pre><code>## will read mTet13_P08</code></pre>
<pre><code>## will read Tet13_0309_P02</code></pre>
<pre><code>## will read Tet13_0309_P03</code></pre>
<pre><code>## will read Tet13_0309_P04</code></pre>
<pre><code>## will read Tet13_0309_P05</code></pre>
<pre><code>## will read Tet13_2_0613</code></pre>
<pre><code>## will read Tet13_DKO_3</code></pre>
<pre><code>## will read Tet13_DKO_4</code></pre>
<pre><code>## will read Tet13_P05</code></pre>
<pre><code>## will read Tet13_P06</code></pre>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="import-tet-23-dko-chimera-embryos" class="section level3 hasAnchor">
<h3><span class="header-section-number">2.0.6</span> Import Tet 2/3 DKO chimera embryos<a href="import-scrna-seq-plates.html#import-tet-23-dko-chimera-embryos" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="import-scrna-seq-plates.html#cb224-1"></a>mat_nm &lt;-<span class="st"> &quot;dko23_chim&quot;</span></span>
<span id="cb224-2"><a href="import-scrna-seq-plates.html#cb224-2"></a></span>
<span id="cb224-3"><a href="import-scrna-seq-plates.html#cb224-3"></a>metadata &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&quot;data/plate_metadata.csv&quot;</span>, <span class="dt">stringsAsFactors =</span> F)</span>
<span id="cb224-4"><a href="import-scrna-seq-plates.html#cb224-4"></a>genotype &lt;-<span class="st"> &quot;Tet2/3 DKO&quot;</span></span>
<span id="cb224-5"><a href="import-scrna-seq-plates.html#cb224-5"></a>treatment &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Chimera assay&quot;</span>)</span>
<span id="cb224-6"><a href="import-scrna-seq-plates.html#cb224-6"></a>metadata &lt;-<span class="st"> </span>metadata[metadata<span class="op">$</span>treatment <span class="op">%in%</span><span class="st"> </span>treatment, ]</span>
<span id="cb224-7"><a href="import-scrna-seq-plates.html#cb224-7"></a>metadata &lt;-<span class="st"> </span>metadata[<span class="kw">grep</span>(<span class="dt">pattern =</span> genotype, <span class="dt">x =</span> metadata<span class="op">$</span>genotype), ]</span>
<span id="cb224-8"><a href="import-scrna-seq-plates.html#cb224-8"></a></span>
<span id="cb224-9"><a href="import-scrna-seq-plates.html#cb224-9"></a><span class="kw">import_plates</span>(<span class="dt">mat_nm =</span> mat_nm, <span class="dt">metadata =</span> metadata)</span></code></pre></div>
<pre><code>## will read Tet23_0303_P01</code></pre>
<pre><code>## will read Tet23_0303_P02</code></pre>
<pre><code>## will read Tet23_0303_P03</code></pre>
<pre><code>## will read Tet23_1</code></pre>
<pre><code>## will read Tet23_2</code></pre>
<pre><code>## will read Tet23_4</code></pre>
<pre><code>## will read Tet23_5</code></pre>
<pre><code>## will read Tet23_7</code></pre>
<pre><code>## will read Tet23_8</code></pre>
<pre><code>## will read Tet23_DKO_1_0616</code></pre>
<pre><code>## will read Tet23_DKO_2_0616</code></pre>
<pre><code>## will read Tet23DKO_3_0616</code></pre>
<pre><code>## will read Tet23DKO_4_0616</code></pre>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="import-aav-cre-delivered-whole-embryo-tkos" class="section level3 hasAnchor">
<h3><span class="header-section-number">2.0.7</span> Import AAV-Cre delivered whole-embryo TKOs<a href="import-scrna-seq-plates.html#import-aav-cre-delivered-whole-embryo-tkos" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="import-scrna-seq-plates.html#cb239-1"></a>mat_nm &lt;-<span class="st"> &quot;tko_germline&quot;</span></span>
<span id="cb239-2"><a href="import-scrna-seq-plates.html#cb239-2"></a></span>
<span id="cb239-3"><a href="import-scrna-seq-plates.html#cb239-3"></a>metadata &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&quot;data/plate_metadata.csv&quot;</span>, <span class="dt">stringsAsFactors =</span> F)</span>
<span id="cb239-4"><a href="import-scrna-seq-plates.html#cb239-4"></a>genotype &lt;-<span class="st"> &quot;TKO&quot;</span></span>
<span id="cb239-5"><a href="import-scrna-seq-plates.html#cb239-5"></a>treatment &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;AAV Cre-delivered Tet1/2/3 triple KO&quot;</span>)</span>
<span id="cb239-6"><a href="import-scrna-seq-plates.html#cb239-6"></a>metadata &lt;-<span class="st"> </span>metadata[metadata<span class="op">$</span>treatment <span class="op">%in%</span><span class="st"> </span>treatment, ]</span>
<span id="cb239-7"><a href="import-scrna-seq-plates.html#cb239-7"></a>metadata &lt;-<span class="st"> </span>metadata[<span class="kw">grep</span>(<span class="dt">pattern =</span> genotype, <span class="dt">x =</span> metadata<span class="op">$</span>genotype), ]</span>
<span id="cb239-8"><a href="import-scrna-seq-plates.html#cb239-8"></a></span>
<span id="cb239-9"><a href="import-scrna-seq-plates.html#cb239-9"></a><span class="kw">import_plates</span>(<span class="dt">mat_nm =</span> mat_nm, <span class="dt">metadata =</span> metadata)</span></code></pre></div>
<pre><code>## will read TetAAV_P01</code></pre>
<pre><code>## will read TetAAV_P07</code></pre>
<pre><code>## will read TetAAV_P08</code></pre>
<pre><code>## will read TetAAV_P10</code></pre>
<pre><code>## will read TetAAV_P11</code></pre>
<pre><code>## will read TetAAV_P12</code></pre>
<pre><code>## will read TetAAV_P16</code></pre>
<pre><code>## will read TetAAV_P18</code></pre>
<pre><code>## will read TetAAV_P13_2</code></pre>
<pre><code>## will read TetAAV_P19_2</code></pre>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="import-scrna-seq-plates.html#cb251-1"></a><span class="co">#### next step: gating of populations</span></span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="facs-gating-of-cell-population.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
